diff --git a/genutil/common.go b/genutil/common.go
index b7c81e3..c8f9100 100644
--- a/genutil/common.go
+++ b/genutil/common.go
@@ -18,7 +18,6 @@ package genutil
 
 import (
 	"fmt"
-	"io"
 	"sort"
 	"strings"
 
@@ -43,7 +42,7 @@ const (
 )
 
 // WriteIfNotEmpty writes the string s to b if it has a non-zero length.
-func WriteIfNotEmpty(b io.StringWriter, s string) {
+func WriteIfNotEmpty(b *strings.Builder, s string) {
 	if len(s) != 0 {
 		b.WriteString(s)
 	}
diff --git a/util/debug.go b/util/debug.go
index ee4c9e4..b25f1b1 100644
--- a/util/debug.go
+++ b/util/debug.go
@@ -12,6 +12,9 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// This file is changed by Broadcom.
+// Modifications - Copyright 2019 Broadcom. The term Broadcom refers to Broadcom Inc. and/or its subsidiaries.
+
 package util
 
 import (
@@ -53,6 +56,14 @@ func DbgPrint(v ...interface{}) {
 	fmt.Println(globalIndent + out)
 }
 
+func IsDebugLibraryEnabled() bool {
+	return debugLibrary
+}
+
+func IsDebugSchemaEnabled() bool {
+	return debugSchema
+}
+
 // DbgSchema prints v if the package global variable debugSchema is set.
 // v has the same format as Printf.
 func DbgSchema(v ...interface{}) {
@@ -177,6 +188,9 @@ func SchemaTypeStr(schema *yang.Entry) string {
 
 // YangTypeToDebugString returns a debug string representation of a YangType.
 func YangTypeToDebugString(yt *yang.YangType) string {
+	if !debugLibrary {
+		return ""
+	}
 	out := fmt.Sprintf("(TypeKind: %s", yang.TypeKindToName[yt.Kind])
 	if p, isPOSIX := SanitizedPattern(yt); len(p) != 0 {
 		out += fmt.Sprintf(", Sanitized pattern (POSIX: %v): %s", isPOSIX, strings.Join(p, " or "))
diff --git a/util/path.go b/util/path.go
index 51e1eb3..5bbfab0 100644
--- a/util/path.go
+++ b/util/path.go
@@ -12,6 +12,9 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// This file is changed by Broadcom.
+// Modifications - Copyright 2019 Broadcom. The term Broadcom refers to Broadcom Inc. and/or its subsidiaries.
+
 package util
 
 import (
@@ -19,10 +22,14 @@ import (
 	"fmt"
 	"reflect"
 	"strings"
+	"sync"
 
 	"github.com/openconfig/goyang/pkg/yang"
 )
 
+var pathToSchemaCache map[reflect.StructTag][]string = make(map[reflect.StructTag][]string)
+var pathToSchemaMutex sync.RWMutex
+
 // SchemaPaths returns all the paths in the path tag.
 func SchemaPaths(f reflect.StructField) ([][]string, error) {
 	var out [][]string
@@ -56,34 +63,52 @@ func ShadowSchemaPaths(f reflect.StructField) [][]string {
 
 // RelativeSchemaPath returns a path to the schema for the struct field f.
 // Paths are embedded in the "path" struct tag and can be either simple:
-//   e.g. "path:a"
+//
+//     e.g. "path:a"
+//
 // or composite (if path compression is used) e.g.
-//   e.g. "path:config/a|a"
+//
+//     e.g. "path:config/a|a"
+//
 // In the latter case, this function returns {"config", "a"}, because only the
 // longer path exists in the data tree and we want the schema for that node.
 // This case is found in OpenConfig leaf-ref cases where the key of a list is a
 // leafref; the schema *yang.Entry for the field is given by
 // schema.Dir["config"].Dir["a"].
 func RelativeSchemaPath(f reflect.StructField) ([]string, error) {
-	pathTag, ok := f.Tag.Lookup("path")
-	if !ok || pathTag == "" {
-		return nil, fmt.Errorf("field %s did not specify a path", f.Name)
-	}
+	pathToSchemaMutex.RLock()
+	if pe, ok := pathToSchemaCache[f.Tag]; ok {
+		pathToSchemaMutex.RUnlock()
+		return pe, nil
+	} else {
+		pathToSchemaMutex.RUnlock()
+		pathTag, ok := f.Tag.Lookup("path")
+		if !ok || pathTag == "" {
+			return nil, fmt.Errorf("field %s did not specify a path", f.Name)
+		}
 
-	paths := strings.Split(pathTag, "|")
-	if len(paths) == 1 {
-		pathTag = strings.TrimPrefix(pathTag, "/")
-		return strings.Split(pathTag, "/"), nil
-	}
-	for _, pv := range paths {
-		pv = strings.TrimPrefix(pv, "/")
-		pe := strings.Split(pv, "/")
-		if len(pe) > 1 {
-			return pe, nil
+		paths := strings.Split(pathTag, "|")
+		if len(paths) == 1 {
+			pathTag = strings.TrimPrefix(pathTag, "/")
+			retPath := strings.Split(pathTag, "/")
+			pathToSchemaMutex.Lock()
+			pathToSchemaCache[f.Tag] = retPath
+			pathToSchemaMutex.Unlock()
+			return retPath, nil
+		}
+		for _, pv := range paths {
+			pv = strings.TrimPrefix(pv, "/")
+			pe := strings.Split(pv, "/")
+			if len(pe) > 1 {
+				pathToSchemaMutex.Lock()
+				pathToSchemaCache[f.Tag] = pe
+				pathToSchemaMutex.Unlock()
+				return pe, nil
+			}
 		}
-	}
 
-	return nil, fmt.Errorf("field %s had path tag %s with |, but no elements of form a/b", f.Name, pathTag)
+		return nil, fmt.Errorf("field %s had path tag %s with |, but no elements of form a/b", f.Name, pathTag)
+	}
 }
 
 // SchemaTreePath returns the schema tree path of the supplied yang.Entry
@@ -200,10 +225,10 @@ func removeXPATHPredicates(s string) (string, error) {
 // FindLeafRefSchema returns a schema Entry at the path pathStr relative to
 // schema if it exists, or an error otherwise.
 // pathStr has either:
-//  - the relative form "../a/b/../b/c", where ".." indicates the parent of the
-//    node, or
-//  - the absolute form "/a/b/c", which indicates the absolute path from the
-//    root of the schema tree.
+//   - the relative form "../a/b/../b/c", where ".." indicates the parent of the
+//     node, or
+//   - the absolute form "/a/b/c", which indicates the absolute path from the
+//     root of the schema tree.
 func FindLeafRefSchema(schema *yang.Entry, pathStr string) (*yang.Entry, error) {
 	if pathStr == "" {
 		return nil, fmt.Errorf("leafref schema %s has empty path", schema.Name)
@@ -246,6 +271,10 @@ func FindLeafRefSchema(schema *yang.Entry, pathStr string) (*yang.Entry, error)
 		refSchema = entries[pe]
 	}
 
+	if refSchema.Type.Kind == yang.Yleafref {
+		return FindLeafRefSchema(refSchema, refSchema.Type.Path)
+	}
+
 	return refSchema, nil
 }
 
diff --git a/util/reflect.go b/util/reflect.go
index 7b64177..7b8c8ad 100644
--- a/util/reflect.go
+++ b/util/reflect.go
@@ -12,6 +12,9 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// This file is changed by Broadcom.
+// Modifications - Copyright 2019 Broadcom. The term Broadcom refers to Broadcom Inc. and/or its subsidiaries.
+
 package util
 
 import (
@@ -282,7 +285,7 @@ func InsertIntoStruct(parentStruct interface{}, fieldName string, fieldValue int
 		n = reflect.Zero(ft.Type)
 	}
 
-	if !isFieldTypeCompatible(ft, n) {
+	if !isFieldTypeCompatible(ft, n) && !IsValueTypeCompatible(ft.Type, v) {
 		return fmt.Errorf("cannot assign value %v (type %T) to struct field %s (type %v) in struct %T", fieldValue, fieldValue, fieldName, ft.Type, parentStruct)
 	}
 
@@ -453,13 +456,45 @@ func DeepEqualDerefPtrs(a, b interface{}) bool {
 	return reflect.DeepEqual(aa, bb)
 }
 
+func updateChildSchemaCache(schema *yang.Entry, tagStr reflect.StructTag, ygEntry *yang.Entry) {
+	schema.PerfOpts.ChildSchemaMutex.Lock()
+	schema.PerfOpts.ChildSchemaCache[tagStr] = ygEntry
+	schema.PerfOpts.ChildSchemaMutex.Unlock()
+}
+
+func childSchemaCache(schema *yang.Entry, tagStr reflect.StructTag) *yang.Entry {
+	if schema.PerfOpts.ChildSchemaCache == nil {
+		schema.PerfOpts.ChildSchemaMutex.Lock()
+		if schema.PerfOpts.ChildSchemaCache == nil {
+			schema.PerfOpts.ChildSchemaCache = make(map[reflect.StructTag]*yang.Entry)
+			schema.PerfOpts.ChildSchemaMutex.Unlock()
+			return nil
+		}
+		schema.PerfOpts.ChildSchemaMutex.Unlock()
+	}
+	schema.PerfOpts.ChildSchemaMutex.RLock()
+	rv, ok := schema.PerfOpts.ChildSchemaCache[tagStr]
+	schema.PerfOpts.ChildSchemaMutex.RUnlock()
+	if ok {
+		return rv
+	}
+	return nil
+}
+
 // ChildSchema returns the schema for the struct field f, if f contains a valid
 // path tag and the schema path is found in the schema tree. It returns an error
 // if the struct tag is invalid, or nil if tag is valid but the schema is not
 // found in the tree at the specified path.
 func ChildSchema(schema *yang.Entry, f reflect.StructField) (*yang.Entry, error) {
-	pathTag, _ := f.Tag.Lookup("path")
-	DbgSchema("childSchema for schema %s, field %s, tag %s\n", schema.Name, f.Name, pathTag)
+	if cschema := childSchemaCache(schema, f.Tag); cschema != nil {
+		return cschema, nil
+	}
+
+	if IsDebugSchemaEnabled() {
+		pathTag, _ := f.Tag.Lookup("path")
+		DbgSchema("childSchema for schema %s, field %s, tag %s\n", schema.Name, f.Name, pathTag)
+	}
+
 	p, err := RelativeSchemaPath(f)
 	if err != nil {
 		return nil, err
@@ -490,6 +525,7 @@ func ChildSchema(schema *yang.Entry, f reflect.StructField) (*yang.Entry, error)
 	}
 	if foundSchema {
 		DbgSchema(" - found\n")
+		updateChildSchemaCache(schema, f.Tag, childSchema)
 		return childSchema, nil
 	}
 	DbgSchema(" - not found\n")
@@ -505,11 +541,15 @@ func ChildSchema(schema *yang.Entry, f reflect.StructField) (*yang.Entry, error)
 		// path element i.e. choice1/case1/leaf1 path in the schema will have
 		// struct tag `path:"leaf1"`. This implies that only paths with length
 		// 1 are eligible for this matching.
+		updateChildSchemaCache(schema, f.Tag, nil)
 		return nil, nil
 	}
 	entries := FindFirstNonChoiceOrCase(schema)
 
-	DbgSchema("checking for %s against non choice/case entries: %v\n", p[0], stringMapKeys(entries))
+	if IsDebugSchemaEnabled() {
+		DbgSchema("checking for %s against non choice/case entries: %v\n", p[0], stringMapKeys(entries))
+	}
+
 	for path, entry := range entries {
 		splitPath := SplitPath(path)
 		name := splitPath[len(splitPath)-1]
@@ -517,11 +557,13 @@ func ChildSchema(schema *yang.Entry, f reflect.StructField) (*yang.Entry, error)
 
 		if StripModulePrefix(name) == p[0] {
 			DbgSchema(" - match\n")
+			updateChildSchemaCache(schema, f.Tag, entry)
 			return entry, nil
 		}
 	}
 
 	DbgSchema(" - no matches\n")
+	updateChildSchemaCache(schema, f.Tag, nil)
 	return nil, nil
 }
 
diff --git a/ygen/codegen.go b/ygen/codegen.go
index c5b09df..6258c8d 100644
--- a/ygen/codegen.go
+++ b/ygen/codegen.go
@@ -15,6 +15,10 @@
 // Package ygen contains a library to generate Go structs from a YANG model.
 // The Goyang parsing library is used to parse YANG. The output can consider
 // OpenConfig-specific conventions such that the schema is compressed.
+
+// This file is changed by Broadcom.
+// Modifications - Copyright 2019 Broadcom. The term Broadcom refers to Broadcom Inc. and/or its subsidiaries.
+
 package ygen
 
 import (
@@ -1082,7 +1086,7 @@ func findMappableEntities(e *yang.Entry, dirs map[string]*yang.Entry, enums map[
 			dirs[ch.Path()] = ch
 			// Recurse down the tree.
 			errs = util.AppendErrs(errs, findMappableEntities(ch, dirs, enums, excludeModules, compressPaths, modules))
-		case ch.Kind == yang.AnyDataEntry:
+		case (ch.Kind == yang.AnyDataEntry), (ch.Kind == yang.NotificationEntry):
 			continue
 		default:
 			errs = util.AppendErr(errs, fmt.Errorf("unknown type of entry %v in findMappableEntities for %s", ch.Kind, ch.Path()))
@@ -1135,6 +1139,7 @@ func MakeFakeRoot(rootName string) *yang.Entry {
 		Node: &yang.Value{
 			Name: rootElementNodeName,
 		},
+		PerfOpts: &yang.PerformanceOpts{},
 	}
 }
 
diff --git a/ygen/genstate.go b/ygen/genstate.go
index ad0bbc7..e308ef0 100644
--- a/ygen/genstate.go
+++ b/ygen/genstate.go
@@ -12,6 +12,9 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// This file is changed by Broadcom.
+// Modifications - Copyright 2019 Broadcom. The term Broadcom refers to Broadcom Inc. and/or its subsidiaries.
+
 package ygen
 
 import (
diff --git a/ygen/schemaparse.go b/ygen/schemaparse.go
index 46c3127..ebc02ff 100644
--- a/ygen/schemaparse.go
+++ b/ygen/schemaparse.go
@@ -40,13 +40,32 @@ func buildJSONTree(ms []*yang.Entry, dn map[string]string, fakeroot *yang.Entry,
 	rootEntry := &yang.Entry{
 		Dir:        map[string]*yang.Entry{},
 		Annotation: map[string]interface{}{},
+		PerfOpts:   &yang.PerformanceOpts{},
 	}
 	for _, m := range ms {
 		annotateChildren(m, dn, inclDescriptions)
+
+		// a leaf with a different prefix
+		if len(m.Dir) == 0 {
+			if _, found := m.Annotation["modulename"]; found {
+				rootEntry.Dir[m.Name] = m
+			}
+		}
+
 		for _, ch := range util.Children(m) {
 			if _, ex := rootEntry.Dir[ch.Name]; ex {
 				return nil, fmt.Errorf("overlapping root children for key %s", ch.Name)
 			}
+
+			ch.Annotation["namespace"] = m.Namespace().Name
+			ch.Annotation["modulename"] = m.Prefix.Parent.NName()
+			ch.Annotation["revision"] = m.Prefix.Parent.(*yang.Module).Current()
+			if m.Prefix.Parent.(*yang.Module).Organization != nil {
+				ch.Annotation["organization"] = m.Prefix.Parent.(*yang.Module).Organization.Name
+			} else {
+				ch.Annotation["organization"] = ""
+			}
+
 			rootEntry.Dir[ch.Name] = ch
 		}
 	}
@@ -102,6 +121,7 @@ func annotateChildren(e *yang.Entry, dn map[string]string, inclDescriptions bool
 func annotateEntry(e *yang.Entry, dn map[string]string, inclDescriptions bool) {
 	if !inclDescriptions {
 		e.Description = ""
+		e.Augmented = nil
 	}
 	if e.Annotation == nil {
 		e.Annotation = map[string]interface{}{}
@@ -112,6 +132,11 @@ func annotateEntry(e *yang.Entry, dn map[string]string, inclDescriptions bool) {
 	if e.IsDir() {
 		e.Annotation["schemapath"] = e.Path()
 	}
+	// add the modulename if the prefix is different from its parent
+	// and it belong to "module" only, i.e. filter out submodule
+	if e.Parent != nil && e.Namespace().Name != e.Parent.Namespace().Name && e.Prefix.Parent.Statement().Keyword == "module" {
+		e.Annotation["modulename"] = e.Prefix.Parent.NName()
+	}
 }
 
 // WriteGzippedByteSlice takes an input slice of bytes, gzips it
diff --git a/ygot/render.go b/ygot/render.go
index 307e579..8f185be 100644
--- a/ygot/render.go
+++ b/ygot/render.go
@@ -20,6 +20,7 @@ import (
 	"fmt"
 	"reflect"
 	"sort"
+	"strconv"
 	"strings"
 
 	"github.com/openconfig/gnmi/errlist"
@@ -88,6 +89,11 @@ type path struct {
 	p *gnmiPath
 }
 
+type structTagInfo struct {
+	gPathList []*gnmiPath
+	chModName string
+}
+
 func (p *path) String() string {
 	if p.p.isPathElemPath() {
 		return prototext.Format(&gnmipb.Path{Elem: p.p.pathElemPath})
@@ -315,6 +321,8 @@ type GNMINotificationsConfig struct {
 	// of PathElem messages. This path format is used by gNMI 0.4.0 and
 	// above. Used if PathElem is set.
 	PathElemPrefix []*gnmipb.PathElem
+	// Encoding to be used for atomic values - like keyless lists
+	Encoding gnmipb.Encoding
 }
 
 // TogNMINotifications takes an input GoStruct and renders it to slice of
@@ -340,7 +348,7 @@ func TogNMINotifications(s GoStruct, ts int64, cfg GNMINotificationsConfig) ([]*
 		return nil, err
 	}
 
-	msgs, err := leavesToNotifications(leaves, ts, pfx)
+	msgs, err := leavesToNotifications(leaves, ts, pfx, cfg.Encoding)
 	if err != nil {
 		return nil, err
 	}
@@ -422,6 +430,13 @@ func findUpdatedLeaves(leaves map[*path]interface{}, s GoStruct, parent *gnmiPat
 			}
 		case reflect.Slice:
 			if fval.Type().Elem().Kind() == reflect.Ptr {
+				// s is a container with single keyless list. Add the whole container
+				if sval.NumField() == 1 {
+					for _, p := range mapPaths {
+						leaves[&path{p}] = s
+					}
+					continue
+				}
 				// This is a keyless list - currently unsupported for mapping since there is
 				// not an explicit path that can be used.
 				errs.Add(fmt.Errorf("unimplemented: keyless list cannot be output: %v", mapPaths[0]))
@@ -629,7 +644,7 @@ func sliceToScalarArray(v []interface{}) (*gnmipb.ScalarArray, error) {
 // likely to be suboptimal since it results in very large Notifications for particular
 // structs. There should be some fragmentation of Updates across Notification messages
 // in a future implementation. We return a slice to keep the API stable.
-func leavesToNotifications(leaves map[*path]interface{}, ts int64, pfx *gnmiPath) ([]*gnmipb.Notification, error) {
+func leavesToNotifications(leaves map[*path]interface{}, ts int64, pfx *gnmiPath, enc gnmipb.Encoding) ([]*gnmipb.Notification, error) {
 	n := &gnmipb.Notification{
 		Timestamp: ts,
 	}
@@ -651,7 +666,7 @@ func leavesToNotifications(leaves map[*path]interface{}, ts int64, pfx *gnmiPath
 			return nil, err
 		}
 
-		val, err := EncodeTypedValue(v, gnmipb.Encoding_JSON)
+		val, err := EncodeTypedValue(v, enc)
 		if err != nil {
 			return nil, err
 		}
@@ -762,7 +777,7 @@ func marshalStruct(s GoStruct, enc gnmipb.Encoding) (*gnmipb.TypedValue, error)
 		return nil, err
 	}
 
-	js, err := json.MarshalIndent(j, "", "  ")
+	js, err := json.Marshal(j)
 	if err != nil {
 		return nil, fmt.Errorf("cannot encode JSON, %v", err)
 	}
@@ -944,19 +959,21 @@ func (*RFC7951JSONConfig) IsMarshal7951Arg() {}
 // to JSON described by RFC7951. The supplied args control options corresponding
 // to the method by which JSON is marshalled.
 func ConstructIETFJSON(s GoStruct, args *RFC7951JSONConfig) (map[string]interface{}, error) {
+	tagPaths := map[reflect.StructTag]*structTagInfo{}
 	return structJSON(s, "", jsonOutputConfig{
 		jType:         RFC7951,
 		rfc7951Config: args,
-	})
+	}, tagPaths)
 }
 
 // ConstructInternalJSON marshals a supplied GoStruct to a map, suitable for handing
 // to json.Marshal. It uses the loosely specified JSON format document in
 // go/yang-internal-json.
 func ConstructInternalJSON(s GoStruct) (map[string]interface{}, error) {
+	tagPaths := map[reflect.StructTag]*structTagInfo{}
 	return structJSON(s, "", jsonOutputConfig{
 		jType: Internal,
-	})
+	}, tagPaths)
 }
 
 // Marshal7951Arg is an interface implemented by arguments to
@@ -993,10 +1010,11 @@ func Marshal7951(d interface{}, args ...Marshal7951Arg) ([]byte, error) {
 			indent = string(v)
 		}
 	}
+	tagPaths := map[reflect.StructTag]*structTagInfo{}
 	j, err := jsonValue(reflect.ValueOf(d), "", jsonOutputConfig{
 		jType:         RFC7951,
 		rfc7951Config: rfcCfg,
-	})
+	}, tagPaths)
 
 	if err != nil {
 		return nil, err
@@ -1094,7 +1112,7 @@ func appmodsJSON(fType reflect.StructField, parentMod string, args jsonOutputCon
 // be produced and whether such module names are appended is controlled through the
 // supplied jsonOutputConfig. Returns an error if the GoStruct cannot be rendered
 // to JSON.
-func structJSON(s GoStruct, parentMod string, args jsonOutputConfig) (map[string]interface{}, error) {
+func structJSON(s GoStruct, parentMod string, args jsonOutputConfig, tagPaths map[reflect.StructTag]*structTagInfo) (map[string]interface{}, error) {
 	var errs errlist.List
 
 	sval := reflect.ValueOf(s).Elem()
@@ -1108,31 +1126,61 @@ func structJSON(s GoStruct, parentMod string, args jsonOutputConfig) (map[string
 		field := sval.Field(i)
 		fType := stype.Field(i)
 
-		// Module names to append to the path in RFC7951 output mode.
-		var appmods [][]string
-		var chMod string
-		if args.jType == RFC7951 && args.rfc7951Config != nil && args.rfc7951Config.AppendModuleName {
-			var err error
-			if appmods, chMod, err = appmodsJSON(fType, parentMod, args); err != nil {
-				errs.Add(err)
+		switch field.Kind() {
+		case reflect.Map, reflect.Slice, reflect.Ptr, reflect.Interface:
+			if field.IsNil() {
 				continue
 			}
 		}
 
-		mapPaths, err := structTagToLibPaths(fType, newStringSliceGNMIPath([]string{}), args.rfc7951Config != nil && args.rfc7951Config.PreferShadowPath)
-		if err != nil {
-			errs.Add(fmt.Errorf("%s: %v", fType.Name, err))
-			continue
+		// Determine whether we should append a module name to the path in RFC7951
+		// output mode.
+		var appmod string
+
+		pmod := parentMod
+
+		var mapPaths []*gnmiPath
+		var err error
+
+		if tagInfo, ok := tagPaths[fType.Tag]; !ok {
+			tagInfoObj := new(structTagInfo)
+			tagInfoObj.gPathList, err = structTagToLibPaths(fType, newStringSliceGNMIPath([]string{}), args.rfc7951Config != nil && args.rfc7951Config.PreferShadowPath)
+			if err != nil {
+				errs.Add(fmt.Errorf("%s: %v", fType.Name, err))
+				continue
+			}
+			if chMod, modOk := fType.Tag.Lookup("module"); modOk {
+				// If the child module isn't the same as the parent module,
+				// then appmod stores the name of the module to prefix to pathsmakejson
+				// within this context.
+				if chMod != parentMod {
+					appmod = chMod
+				}
+				// Update the parent module name to be used for subsequent
+				// children.
+				pmod = chMod
+				tagInfoObj.chModName = chMod
+			} else {
+				tagInfoObj.chModName = pmod
+			}
+			tagPaths[fType.Tag] = tagInfoObj
+			mapPaths = tagInfoObj.gPathList
+		} else {
+			mapPaths = tagInfo.gPathList
+			pmod = tagInfo.chModName
+			if tagInfo.chModName != parentMod {
+				appmod = tagInfo.chModName
+			}
 		}
 
-		// s is the fake root if its path tag is empty. In this case,
-		// we want to forward the parent module to the child nodes.
-		isFakeRoot := len(mapPaths) == 1 && mapPaths[0].Len() == 0
-		if isFakeRoot {
-			chMod = parentMod
+		var appendModName bool
+		if args.jType == RFC7951 && args.rfc7951Config != nil && args.rfc7951Config.AppendModuleName && appmod != "" {
+			appendModName = true
 		}
 
-		value, err := jsonValue(field, chMod, args)
+		var value interface{}
+		value, err = jsonValue(field, pmod, args, tagPaths)
+
 		if err != nil {
 			errs.Add(err)
 			continue
@@ -1146,55 +1194,85 @@ func structJSON(s GoStruct, parentMod string, args jsonOutputConfig) (map[string
 			continue
 		}
 
-		if isFakeRoot {
-			if v, ok := value.(map[string]interface{}); ok {
-				for mk, mv := range v {
-					jsonout[mk] = mv
+		for _, p := range mapPaths {
+			v, ok := value.(map[string]interface{})
+			switch p.Len() {
+			case 0:
+				if ok {
+					for mk, mv := range v {
+						k := mk
+						if appendModName {
+							// Append the module name for the 0th element if
+							// specified to do so. This is the module name of
+							// the root.
+							k = fmt.Sprintf("%s:%s", appmod, mk)
+						}
+						jsonout[k] = mv
+					}
+				} else {
+					errs.Add(fmt.Errorf("empty path specified for non-root entity"))
+					continue
 				}
-			} else {
-				errs.Add(fmt.Errorf("empty path specified for non-root entity"))
-			}
-			continue
-		}
-
-		if appmods != nil && len(mapPaths) != len(appmods) {
-			errs.Add(fmt.Errorf("%s: number of paths and modules in struct tag not the same: (paths: %v, modules: %v)", fType.Name, len(mapPaths), len(appmods)))
-			continue
-		}
-
-		for i, p := range mapPaths {
-			if appmods != nil && p.Len() != len(appmods[i]) {
-				errs.Add(fmt.Errorf("number of paths and modules elements not the same: (paths: %v, modules: %v)", p, appmods[i]))
-				continue
-			}
-
-			parent := jsonout
-			j := 0
-			for ; j != p.Len()-1; j++ {
-				k, err := p.StringElemAt(j)
+			case 1:
+				pelem, err := p.StringElemAt(0)
 				if err != nil {
 					errs.Add(err)
 					continue
 				}
-
-				if appmods != nil && appmods[i][j] != "" {
-					k = fmt.Sprintf("%s:%s", appmods[i][j], k)
+				if appendModName {
+					pelem = fmt.Sprintf("%s:%s", appmod, pelem)
 				}
+				jsonout[pelem] = value
+			default:
+				var nilParent bool
+				parent := jsonout
+				for i := 0; i < p.Len()-1; i++ {
+					k, err := p.StringElemAt(i)
+					if err != nil {
+						errs.Add(err)
+						continue
+					}
 
-				if _, ok := parent[k]; !ok {
-					parent[k] = map[string]interface{}{}
+					switch {
+					case (i == 0 && appendModName && !p.isAbsolute):
+						// If the path is not absolute, then path compression has
+						// occurred - and therefore the elements must be in the
+						// same module. In this case, we append the module name
+						// to the first element in the list.
+						fallthrough
+					case i == 0 && appendModName && parentMod == "":
+						// For the first element, regardless of whether the path
+						// was absolute or not, we always must append the module
+						// name if there was no parent module, since this is an
+						// entity at the root.
+						k = fmt.Sprintf("%s:%s", appmod, k)
+						nilParent = true
+					}
+					if _, ok := parent[k]; !ok {
+						parent[k] = map[string]interface{}{}
+					}
+					parent = parent[k].(map[string]interface{})
 				}
-				parent = parent[k].(map[string]interface{})
-			}
-			k, err := p.LastStringElem()
-			if err != nil {
-				errs.Add(err)
-				continue
-			}
-			if appmods != nil && appmods[i][j] != "" {
-				k = fmt.Sprintf("%s:%s", appmods[i][j], k)
+				k, err := p.LastStringElem()
+				if err != nil {
+					errs.Add(err)
+					continue
+				}
+				if p.isAbsolute && appendModName && !nilParent {
+					// If the path was not absolute, then we need to prepend the
+					// module name since the last entity in the path was in a
+					// different module to its parent. We do not need to check the
+					// values of the module names, since in the case that the
+					// module is the same appendModName is false.
+					//
+					// In the case that the parent was nil, then we must not
+					// append the name here, since we must be within the same
+					// module.
+					k = fmt.Sprintf("%s:%s", appmod, k)
+				}
+				parent[k] = value
+
 			}
-			parent[k] = value
 		}
 	}
 
@@ -1210,7 +1288,10 @@ func structJSON(s GoStruct, parentMod string, args jsonOutputConfig) (map[string
 // and float64 values are represented as strings.
 func writeIETFScalarJSON(i interface{}) interface{} {
 	switch reflect.ValueOf(i).Kind() {
-	case reflect.Uint64, reflect.Int64, reflect.Float64:
+	case reflect.Float64:
+		val, _ := i.(float64)
+		return strconv.FormatFloat(val, 'f', -1, 64)
+	case reflect.Uint64, reflect.Int64:
 		return fmt.Sprintf("%v", i)
 	}
 	return i
@@ -1241,7 +1322,7 @@ func keyValue(v reflect.Value, appendModuleName bool) (interface{}, error) {
 // constructs the representation for JSON marshalling that corresponds to it.
 // The module within which the map is defined is specified by the parentMod
 // argument.
-func mapJSON(field reflect.Value, parentMod string, args jsonOutputConfig) (interface{}, error) {
+func mapJSON(field reflect.Value, parentMod string, args jsonOutputConfig, tagPaths map[reflect.StructTag]*structTagInfo) (interface{}, error) {
 	var errs errlist.List
 	mapKeyMap := map[string]reflect.Value{}
 	// Order of elements determines the order in which keys will be processed.
@@ -1327,7 +1408,7 @@ func mapJSON(field reflect.Value, parentMod string, args jsonOutputConfig) (inte
 			continue
 		}
 
-		val, err := structJSON(goStruct, parentMod, args)
+		val, err := structJSON(goStruct, parentMod, args, tagPaths)
 		if err != nil {
 			errs.Add(err)
 			continue
@@ -1355,7 +1436,7 @@ func mapJSON(field reflect.Value, parentMod string, args jsonOutputConfig) (inte
 // The module within which the value is defined is specified by the parentMod string,
 // and the type of JSON to be rendered controlled by the value of the jsonOutputConfig
 // provided. Returns an error if one occurs during the mapping process.
-func jsonValue(field reflect.Value, parentMod string, args jsonOutputConfig) (interface{}, error) {
+func jsonValue(field reflect.Value, parentMod string, args jsonOutputConfig, tagPaths map[reflect.StructTag]*structTagInfo) (interface{}, error) {
 	var value interface{}
 	var errs errlist.List
 
@@ -1374,7 +1455,7 @@ func jsonValue(field reflect.Value, parentMod string, args jsonOutputConfig) (in
 	switch field.Kind() {
 	case reflect.Map:
 		var err error
-		value, err = mapJSON(field, parentMod, args)
+		value, err = mapJSON(field, parentMod, args, tagPaths)
 		if err != nil {
 			errs.Add(err)
 		}
@@ -1387,7 +1468,7 @@ func jsonValue(field reflect.Value, parentMod string, args jsonOutputConfig) (in
 			}
 
 			var err error
-			value, err = structJSON(goStruct, parentMod, args)
+			value, err = structJSON(goStruct, parentMod, args, tagPaths)
 			if err != nil {
 				errs.Add(err)
 			}
@@ -1409,7 +1490,7 @@ func jsonValue(field reflect.Value, parentMod string, args jsonOutputConfig) (in
 		case isAnnotationSlice(field):
 			value, err = jsonAnnotationSlice(field)
 		default:
-			value, err = jsonSlice(field, parentMod, args)
+			value, err = jsonSlice(field, parentMod, args, tagPaths)
 		}
 		if err != nil {
 			return nil, err
@@ -1438,13 +1519,13 @@ func jsonValue(field reflect.Value, parentMod string, args jsonOutputConfig) (in
 				return nil, err
 			}
 			if value != nil && reflect.TypeOf(value).Name() == BinaryTypeName {
-				if value, err = jsonSlice(reflect.ValueOf(value), parentMod, args); err != nil {
+				if value, err = jsonSlice(reflect.ValueOf(value), parentMod, args, tagPaths); err != nil {
 					return nil, err
 				}
 				return value, nil
 			}
 		case field.Elem().Kind() == reflect.Slice && field.Elem().Type().Name() == BinaryTypeName:
-			if value, err = jsonSlice(field.Elem(), parentMod, args); err != nil {
+			if value, err = jsonSlice(field.Elem(), parentMod, args, tagPaths); err != nil {
 				return nil, err
 			}
 			return value, nil
@@ -1481,7 +1562,7 @@ func jsonValue(field reflect.Value, parentMod string, args jsonOutputConfig) (in
 // GoStruct, a slice may be a binary field, leaf-list or an unkeyed list. The
 // parentMod is used to track the name of the parent module in the case that
 // module names should be appended.
-func jsonSlice(field reflect.Value, parentMod string, args jsonOutputConfig) (interface{}, error) {
+func jsonSlice(field reflect.Value, parentMod string, args jsonOutputConfig, tagPaths map[reflect.StructTag]*structTagInfo) (interface{}, error) {
 	if field.Type().Name() == BinaryTypeName {
 		// Handle the case that that we have a Binary ([]byte) value,
 		// which must be returned as a JSON string.
@@ -1497,7 +1578,7 @@ func jsonSlice(field reflect.Value, parentMod string, args jsonOutputConfig) (in
 			if !ok {
 				return nil, fmt.Errorf("invalid member of a slice, %s was not a valid GoStruct", c.Name())
 			}
-			j, err := structJSON(gs, parentMod, args)
+			j, err := structJSON(gs, parentMod, args, tagPaths)
 			if err != nil {
 				return nil, err
 			}
@@ -1646,3 +1727,217 @@ func resolveUnionVal(v interface{}, appendModuleName bool) (interface{}, error)
 	}
 	return v, nil
 }
+
+// YGLeafMeta stores the metadata of a leaf value.
+type YGLeafMeta struct {
+        TimestampNano int64
+}
+
+// YGLeafMetaMap defines a map of leaf value to LeafMeta.
+type YGLeafMetaMap map[interface{}]*YGLeafMeta
+
+// YGStructMeta stores the metadata of a ygot.GoStruct object.
+type YGStructMeta struct {
+        TimestampNano int64
+        LeafMeta      YGLeafMetaMap
+}
+
+// YGStructMetaMap defines a map of GoStruct to YGStructMeta.
+type YGStructMetaMap map[GoStruct]*YGStructMeta
+
+// getLeafTimestamp returns leaf's specific timestamp if found in metadata, 0 otherwise.
+func getLeafTimestamp(m *YGStructMeta, leaf interface{}) int64 {
+        if m == nil {
+                return 0
+        }
+
+        ts := m.TimestampNano
+        if m.LeafMeta != nil && len(m.LeafMeta) > 0 {
+                if lm, ok := m.LeafMeta[leaf]; ok {
+                        ts = lm.TimestampNano
+                }
+        }
+
+        if ts > 0 {
+                return ts
+        }
+        return 0
+}
+
+// TogNMINotificationsWithMeta extends ygot.TogNMINotifications to render
+// notifications using meta information stored in metaMap.
+// For example, ygot.TogNMINotifications returns only one notification,
+// TogNMINotificationsWithMeta may split the notification to multiple
+// notifications (with different timestamps) if there are node specific
+// TimestampNano in metaMap.
+func TogNMINotificationsWithMeta(s GoStruct, defaultTs int64, cfg GNMINotificationsConfig, metaMap YGStructMetaMap) ([]*gnmipb.Notification, error) {
+
+        if metaMap == nil || len(metaMap) == 0 {
+                return TogNMINotifications(s, defaultTs, cfg)
+        }
+
+        var pfx *gnmiPath
+        if cfg.UsePathElem {
+                pfx = newPathElemGNMIPath(cfg.PathElemPrefix)
+        } else {
+                pfx = newStringSliceGNMIPath(cfg.StringSlicePrefix)
+        }
+
+        // Assume metaMap is small. We can initialize the tsLeaves map to improve performance.
+        tsLeaves := map[int64]map[*path]interface{}{0: make(map[*path]interface{})}
+        for _, m := range metaMap {
+                if _, ok := tsLeaves[m.TimestampNano]; !ok {
+                        tsLeaves[m.TimestampNano] = make(map[*path]interface{})
+                }
+        }
+
+        // Traverse s and collect all updated leaf nodes into tsLeaves. The updated leaves
+        // are grouped by their timestamps.
+        if err := findUpdatedLeavesWithMeta(tsLeaves, s, pfx, metaMap); err != nil {
+                return nil, err
+        }
+
+        notifications := []*gnmipb.Notification{}
+        for ts, leaves := range tsLeaves {
+                if len(leaves) == 0 {
+                        continue
+                }
+                if ts == 0 {
+                        ts = defaultTs
+                }
+                msgs, err := leavesToNotifications(leaves, ts, pfx, cfg.Encoding)
+                if err != nil {
+                        return nil, err
+                }
+                notifications = append(notifications, msgs...)
+        }
+
+        return notifications, nil
+}
+
+// findUpdatedLeavesWith appends the valid leaves that are within the supplied
+// GoStruct (assumed to the rooted at parentPath) to the supplied tsLeaves map.
+// tsLeaves are grouped by the leaves' timestamps (in supplied YGStructMetaMap).
+// If errors are encountered they are appended to the errlist.List supplied. If
+// the GoStruct contains fields that are themselves structured objects (YANG
+// lists, or containers - represented as maps or struct pointers), the function
+// is called recursively on them.
+func findUpdatedLeavesWithMeta(tsLeaves map[int64]map[*path]interface{}, s GoStruct, parent *gnmiPath, metaMap YGStructMetaMap) error {
+        var errs errlist.List
+
+        if !parent.isValid() {
+                return fmt.Errorf("invalid parent specified: %v", parent)
+        }
+
+        sval := reflect.ValueOf(s)
+        if s == nil || util.IsValueNil(sval) || !sval.IsValid() || !util.IsValueStructPtr(sval) {
+                errs.Add(fmt.Errorf("input struct for %v was not valid", parent))
+                return errs.Err()
+        }
+        sval = sval.Elem()
+
+        stype := sval.Type()
+
+        for i := 0; i < sval.NumField(); i++ {
+                fval := sval.Field(i)
+                ftype := stype.Field(i)
+
+                // Handle nil values, and enumerations specifically.
+                switch fval.Kind() {
+                case reflect.Map, reflect.Slice, reflect.Ptr, reflect.Interface:
+                        if fval.IsNil() {
+                                continue
+                        }
+                }
+
+                mapPaths, err := structTagToLibPaths(ftype, parent, false)
+                if err != nil {
+                        errs.Add(fmt.Errorf("%v->%s: %v", parent, ftype.Name, err))
+                        continue
+                }
+
+                switch fval.Kind() {
+                case reflect.Map:
+                        // We need to map each child along with its key value.
+                        for _, k := range fval.MapKeys() {
+                                childPath, err := mapValuePath(k, fval.MapIndex(k), mapPaths[0])
+                                if err != nil {
+                                        errs.Add(err)
+                                        continue
+                                }
+
+                                goStruct, ok := fval.MapIndex(k).Interface().(GoStruct)
+                                if !ok {
+                                        errs.Add(fmt.Errorf("%v: was not a valid GoStruct", mapPaths[0]))
+                                        continue
+                                }
+                                errs.Add(findUpdatedLeavesWithMeta(tsLeaves, goStruct, childPath, metaMap))
+                        }
+                case reflect.Ptr:
+                        // Determine whether this is a pointer to a struct (another YANG container), or a leaf.
+                        switch fval.Elem().Kind() {
+                        case reflect.Struct:
+                                goStruct, ok := fval.Interface().(GoStruct)
+                                if !ok {
+                                        errs.Add(fmt.Errorf("%v: was not a valid GoStruct", mapPaths[0]))
+                                        continue
+                                }
+                                errs.Add(findUpdatedLeavesWithMeta(tsLeaves, goStruct, mapPaths[0], metaMap))
+                        default:
+                                m := metaMap[s]
+                                for _, p := range mapPaths {
+                                        val := fval.Interface()
+                                        ts := getLeafTimestamp(m, val)
+                                        tsLeaves[ts][&path{p}] = val
+                                }
+                        }
+                case reflect.Slice:
+                        if fval.Type().Elem().Kind() == reflect.Ptr {
+                                // This is a keyless list - currently unsupported for mapping since there is
+                                // not an explicit path that can be used.
+                                errs.Add(fmt.Errorf("unimplemented: keyless list cannot be output: %v", mapPaths[0]))
+                                continue
+                        }
+                        // This is a leaf-list, so add it as though it were a leaf.
+                        m := metaMap[s]
+                        for _, p := range mapPaths {
+                                val := fval.Interface()
+                                ts := getLeafTimestamp(m, val)
+                                tsLeaves[ts][&path{p}] = val
+                        }
+                case reflect.Int64:
+                        name, set, err := enumFieldToString(fval, false)
+                        if err != nil {
+                                errs.Add(err)
+                                continue
+                        }
+
+                        // Skip if the enum has not been explicitly set in the schema.
+                        if !set {
+                                continue
+                        }
+
+                        m := metaMap[s]
+                        ts := getLeafTimestamp(m, fval.Interface())
+                        for _, p := range mapPaths {
+                                tsLeaves[ts][&path{p}] = name
+                        }
+                        continue
+                case reflect.Interface:
+                        // This is a union value.
+                        /*val, err := unionInterfaceValue(fval, false)
+                        if err != nil {
+                                errs.Add(err)
+                                continue
+                        }*/
+                        m := metaMap[s]
+                        for _, p := range mapPaths {
+                                val := fval.Interface()
+                                ts := getLeafTimestamp(m, val)
+                                tsLeaves[ts][&path{p}] = val
+                        }
+                        continue
+                }
+        }
+        return errs.Err()
+}
diff --git a/ygot/struct_validation_map.go b/ygot/struct_validation_map.go
index fcf25b0..7631a8b 100644
--- a/ygot/struct_validation_map.go
+++ b/ygot/struct_validation_map.go
@@ -19,6 +19,10 @@
 //	  to return pointers to a type.
 //	- Renders structs to other output formats such as JSON, or gNMI
 //	  notifications.
+
+// This file is changed by Broadcom.
+// Modifications - Copyright 2019 Broadcom. The term Broadcom refers to Broadcom Inc. and/or its subsidiaries.
+
 package ygot
 
 import (
@@ -401,22 +405,6 @@ type EmitJSONConfig struct {
 // EmitJSON takes an input ValidatedGoStruct (produced by ygen with validation enabled)
 // and serialises it to a JSON string. By default, produces the Internal format JSON.
 func EmitJSON(s ValidatedGoStruct, opts *EmitJSONConfig) (string, error) {
-	var (
-		vopts          []ValidationOption
-		skipValidation bool
-	)
-
-	if opts != nil {
-		vopts = opts.ValidationOpts
-		skipValidation = opts.SkipValidation
-	}
-
-	if !skipValidation {
-		if err := s.Validate(vopts...); err != nil {
-			return "", fmt.Errorf("validation err: %v", err)
-		}
-	}
-
 	v, err := makeJSON(s, opts)
 	if err != nil {
 		return "", err
@@ -424,16 +412,10 @@ func EmitJSON(s ValidatedGoStruct, opts *EmitJSONConfig) (string, error) {
 
 	sb := &strings.Builder{}
 	enc := json.NewEncoder(sb)
-	indent := indentString
 	enc.SetEscapeHTML(false)
 	if opts != nil {
 		enc.SetEscapeHTML(opts.EscapeHTML)
-
-		if opts.Indent != "" {
-			indent = opts.Indent
-		}
 	}
-	enc.SetIndent("", indent)
 
 	if err := enc.Encode(v); err != nil {
 		return "", fmt.Errorf("JSON marshalling error: %v", err)
@@ -633,7 +615,6 @@ func copyStruct(dstVal, srcVal reflect.Value, opts ...MergeOpt) error {
 	for i := 0; i < srcVal.NumField(); i++ {
 		srcField := srcVal.Field(i)
 		dstField := dstVal.Field(i)
-
 		switch srcField.Kind() {
 		case reflect.Ptr:
 			if err := copyPtrField(dstField, srcField, opts...); err != nil {
@@ -666,7 +647,13 @@ func copyStruct(dstVal, srcVal reflect.Value, opts ...MergeOpt) error {
 				dstField.Set(srcField)
 			}
 		default:
-			dstField.Set(srcField)
+			if srcField.Type().Implements(reflect.TypeOf((*GoEnum)(nil)).Elem()) == true {
+				if srcField.Int() != 0 {
+					dstField.Set(srcField)
+				}
+			} else {
+				dstField.Set(srcField)
+			}
 		}
 	}
 	return nil
@@ -925,3 +912,12 @@ func uniqueSlices(a, b reflect.Value) (bool, error) {
 	}
 	return true, nil
 }
+
+func compareMapKeys(existingKeys map[interface{}]reflect.Value, searchKey interface{}) (*reflect.Value, error) {
+	for tmpKey, tmpVal := range existingKeys {
+		if cmp.Equal(tmpKey, searchKey) {
+			return &tmpVal, nil
+		}
+	}
+	return nil, fmt.Errorf("No match found in the existingKeys map.")
+}
diff --git a/ytypes/container.go b/ytypes/container.go
index 7a303d7..5f8de21 100644
--- a/ytypes/container.go
+++ b/ytypes/container.go
@@ -12,6 +12,9 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// This file is changed by Broadcom.
+// Modifications - Copyright 2019 Broadcom. The term Broadcom refers to Broadcom Inc. and/or its subsidiaries.
+
 package ytypes
 
 import (
@@ -71,9 +74,13 @@ func validateContainer(schema *yang.Entry, value ygot.GoStruct) util.Errors {
 			case cschema != nil:
 				// Regular named child.
 				if errs := Validate(cschema, fieldValue); errs != nil {
-					errors = util.AppendErrs(errors, util.PrefixErrors(errs, cschema.Path()))
+					if errs.Error() != "ERROR_READONLY_OBJECT_FOUND" {
+						errors = util.AppendErrs(errors, util.PrefixErrors(errs, cschema.Path()))
+					} else if len(errors) == 0 {
+						errors = util.AppendErrs(errors, errs)
+					}
 				}
-			case !util.IsValueNilOrDefault(structElems.Field(i).Interface()):
+			case !structElems.Field(i).IsNil():
 				// Either an element in choice schema subtree, or bad field.
 				// If the former, it will be found in the choice check below.
 				extraFields[fieldName] = nil
@@ -103,6 +110,10 @@ func validateContainer(schema *yang.Entry, value ygot.GoStruct) util.Errors {
 		errors = util.AppendErr(errors, fmt.Errorf("fields %v are not found in the container schema %s", stringMapSetToSlice(extraFields), schema.Name))
 	}
 
+	if len(errors) == 0 && schema.ReadOnly() == true {
+		errors = util.AppendErrs(errors, util.NewErrs(fmt.Errorf("ERROR_READONLY_OBJECT_FOUND")))
+	}
+
 	return util.UniqueErrors(errors)
 }
 
@@ -140,6 +151,94 @@ func unmarshalContainer(schema *yang.Entry, parent interface{}, jsonTree interfa
 	return unmarshalStruct(schema, parent, jt, enc, opts...)
 }
 
+// unmarshalKeyFieldsInStruct unmarshals a JSON tree into a struct only for the Key fields.
+// schema is the YANG schema of the node corresponding to the struct being unmarshalled into.
+// mapParent is the interface which of type map holds this parent struct with the key
+// parent is the parent struct, which must be a struct ptr.
+// jsonTree is a JSON data tree which must be a map[string]interface{}.
+func unmarshalKeyFieldsInStruct(schema *yang.Entry, mapParent interface{}, parent interface{}, jsonTree map[string]interface{}, enc Encoding, opts ...UnmarshalOpt) (string, error) {
+	destv := reflect.ValueOf(parent).Elem()
+
+	keySet := make(map[string]bool)
+	listKeyType := reflect.TypeOf(mapParent).Key()
+	if util.IsTypeStruct(listKeyType) {
+		newKey := reflect.New(listKeyType).Elem()
+		for i := 0; i < newKey.NumField(); i++ {
+			keySet[listKeyType.Field(i).Name] = true
+		}
+	} else {
+		if fName, err := schemaNameToFieldName(destv, schema.Key); err == nil {
+			keySet[fName] = true
+		}
+	}
+
+	keyStr := ""
+	// Range over the parent struct fields. For each field, check if the data
+	// is present in the JSON tree and if so unmarshal it into the field.
+	for i := 0; i < destv.NumField(); i++ {
+		f := destv.Field(i)
+		ft := destv.Type().Field(i)
+		if _, ok := keySet[ft.Name]; !ok {
+			continue
+		}
+		// Skip annotation fields since they do not have a schema.
+		// TODO(robjs): Implement unmarshalling annotations.
+		if util.IsYgotAnnotation(ft) {
+			continue
+		}
+
+		cschema, err := util.ChildSchema(schema, ft)
+		if err != nil {
+			return "", err
+		}
+		if cschema == nil {
+			return "", fmt.Errorf("unmarshalContainer could not find schema for type %T, field name %s", parent, ft.Name)
+		}
+		jsonValue, err := getJSONTreeValForField(schema, cschema, ft, jsonTree)
+		if err != nil {
+			return "", err
+		}
+		// Store the data tree path of the current field. These will be used
+		// at the end to ensure that there are no excess elements in the JSON
+		// tree not covered by any data path.
+		sp, err := dataTreePaths(schema, cschema, ft)
+		if err != nil {
+			return "", err
+		}
+		if jsonValue == nil {
+			util.DbgPrint("field %s paths %v not present in tree", ft.Name, sp)
+			continue
+		}
+
+		keyStr = keyStr + fmt.Sprintf("%v", jsonValue) + "#"
+		util.DbgPrint("populating field %s type %s with paths %v.", ft.Name, ft.Type, sp)
+		// Only create a new field if it is nil, otherwise update just the
+		// fields that are in the data tree being passed to unmarshal, and
+		// preserve all other existing values.
+		if util.IsNilOrInvalidValue(f) {
+			makeField(destv, ft)
+		}
+
+		p := parent
+		switch {
+		case util.IsUnkeyedList(cschema):
+			// For unkeyed list, we must pass in the addr of the slice to be
+			// able to append to it.
+			p = f.Addr().Interface()
+		case cschema.IsContainer() || cschema.IsList():
+			// For list and container, the new parent is the field we just
+			// created. For leaf and leaf-list, the parent is still the
+			// current container.
+			p = f.Interface()
+		}
+		if err := unmarshalGeneric(cschema, p, jsonValue, enc, opts...); err != nil {
+			return "", err
+		}
+	}
+
+	return keyStr, nil
+}
+
 // unmarshalStruct unmarshals a JSON tree into a struct.
 //   schema is the YANG schema of the node corresponding to the struct being
 //     unmarshalled into.
@@ -248,7 +347,9 @@ func unmarshalStruct(schema *yang.Entry, parent interface{}, jsonTree map[string
 		}
 	}
 
-	util.DbgPrint("container after unmarshal:\n%s\n", pretty.Sprint(destv.Interface()))
+	if util.IsDebugLibraryEnabled() {
+		util.DbgPrint("container after unmarshal:\n%s\n", pretty.Sprint(destv.Interface()))
+	}
 	return nil
 }
 
diff --git a/ytypes/leaf.go b/ytypes/leaf.go
index a6a2617..9975c84 100644
--- a/ytypes/leaf.go
+++ b/ytypes/leaf.go
@@ -12,6 +12,9 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// This file is changed by Broadcom.
+// Modifications - Copyright 2019 Broadcom. The term Broadcom refers to Broadcom Inc. and/or its subsidiaries.
+
 package ytypes
 
 import (
@@ -20,8 +23,10 @@ import (
 	"math/big"
 	"reflect"
 	"strconv"
+	"strings"
 
 	log "github.com/golang/glog"
+	"github.com/openconfig/gnmi/value"
 	"github.com/openconfig/goyang/pkg/yang"
 	"github.com/openconfig/ygot/util"
 	"github.com/openconfig/ygot/ygot"
@@ -252,7 +257,7 @@ func validateUnion(schema *yang.Entry, value interface{}) util.Errors {
 // during validation against each matching schema otherwise.
 func validateMatchingSchemas(schema *yang.Entry, value interface{}) util.Errors {
 	var errors []error
-	ss := findMatchingSchemasInUnion(schema.Type, value)
+	ss := findMatchingSchemasInUnion(schema, schema.Type, value)
 	var kk []yang.TypeKind
 	for _, s := range ss {
 		kk = append(kk, s.Type.Kind)
@@ -283,17 +288,25 @@ func validateMatchingSchemas(schema *yang.Entry, value interface{}) util.Errors
 // findMatchingSchemasInUnion returns all schemas in the given union type,
 // including those within nested unions, that match the Go type of value.
 // value must not be nil.
-func findMatchingSchemasInUnion(ytype *yang.YangType, value interface{}) []*yang.Entry {
+func findMatchingSchemasInUnion(schema *yang.Entry, ytype *yang.YangType, value interface{}) []*yang.Entry {
 	var matches []*yang.Entry
 
 	util.DbgPrint("findMatchingSchemasInUnion for type %T, kind %s", value, reflect.TypeOf(value).Kind())
 	for _, t := range ytype.Type {
 		if t.Kind == yang.Yunion {
 			// Recursively check all union types within this union.
-			matches = append(matches, findMatchingSchemasInUnion(t, value)...)
+			matches = append(matches, findMatchingSchemasInUnion(schema, t, value)...)
 			continue
 		}
 
+		if t.Kind == yang.Yleafref {
+			ns, err := util.FindLeafRefSchema(schema, t.Path)
+			if err != nil {
+				log.Warningf("not found base Go type for type %v in union value %s", t.Kind, util.ValueStr(value))
+				continue
+			}
+			t = ns.Type
+		}
 		ybt := yangBuiltinTypeToGoType(t.Kind)
 		if reflect.ValueOf(value).Kind() == reflect.Ptr {
 			ybt = ygot.ToPtr(yangBuiltinTypeToGoType(t.Kind))
@@ -445,7 +458,10 @@ func unmarshalUnion(schema *yang.Entry, parent interface{}, fieldName string, va
 		return err
 	}
 	if loneType != yang.Ynone {
-		goValue, err := unmarshalScalar(parent, yangKindToLeafEntry(loneType), fieldName, value, enc)
+		ygEntry := yangKindToLeafEntry(loneType)
+		ygEntry.Name = schema.Name
+		ygEntry.Parent = schema.Parent
+		goValue, err := unmarshalScalar(parent, ygEntry, fieldName, value, enc)
 		if err != nil {
 			return fmt.Errorf("could not unmarshal %v into type %s", value, loneType)
 		}
@@ -506,6 +522,8 @@ func unmarshalUnion(schema *yang.Entry, parent interface{}, fieldName string, va
 	for _, sk := range sks {
 		util.DbgPrint("try to unmarshal into type %s", sk)
 		sch := yangKindToLeafEntry(sk)
+		sch.Parent = schema.Parent
+		sch.Name = schema.Name
 		gv, err := unmarshalScalar(parent, sch, fieldName, value, enc)
 		if err == nil {
 			return setUnionFieldWithTypedValue(parentT, destUnionFieldV, destUnionFieldElemT, gv)
@@ -615,22 +633,24 @@ func getUnionTypesNotEnums(schema *yang.Entry, yt *yang.YangType) ([]*yang.YangT
 // type) for a given schema, which must be for an enum type. t is the type of
 // the containing parent struct.
 func schemaToEnumTypes(schema *yang.Entry, t reflect.Type) ([]reflect.Type, error) {
-	enumTypesMethod := reflect.New(t).Elem().MethodByName("ΛEnumTypeMap")
+	enumTypesMethod := reflect.New(t).Elem().MethodByName("\u039bEnumTypeMap")
 	if !enumTypesMethod.IsValid() {
-		return nil, fmt.Errorf("type %s does not have a ΛEnumTypesMap function", t)
+		return nil, fmt.Errorf("type %s does not have a \u039bEnumTypesMap function", t)
 	}
 
 	ec := enumTypesMethod.Call(nil)
 	if len(ec) == 0 {
-		return nil, fmt.Errorf("%s ΛEnumTypes function returns empty value", t)
+		return nil, fmt.Errorf("%s \u039bEnumTypes function returns empty value", t)
 	}
 	ei := ec[0].Interface()
 	enumTypesMap, ok := ei.(map[string][]reflect.Type)
 	if !ok {
-		return nil, fmt.Errorf("%s ΛEnumTypes function returned wrong type %T, want map[string][]reflect.Type", t, ei)
+		return nil, fmt.Errorf("%s \u039bEnumTypes function returned wrong type %T, want map[string][]reflect.Type", t, ei)
 	}
 
-	util.DbgPrint("path is %s for schema %s", absoluteSchemaDataPath(schema), schema.Name)
+	if util.IsDebugLibraryEnabled() {
+		util.DbgPrint("path is %s for schema %s", absoluteSchemaDataPath(schema), schema.Name)
+	}
 
 	return enumTypesMap[absoluteSchemaDataPath(schema)], nil
 }
@@ -756,6 +776,65 @@ func sanitizeJSON(parent interface{}, schema *yang.Entry, fieldName string, valu
 	return nil, fmt.Errorf("unmarshalScalar: unsupported type %v in schema node %s", ykind, schema.Name)
 }
 
+func yangTypeToGoTypeVal(parent interface{}, schema *yang.Entry, fieldName string, value interface{}) (interface{}, error) {
+	ykind := schema.Type.Kind
+
+	switch ykind {
+	case yang.Ybinary:
+		v, err := base64.StdEncoding.DecodeString(value.(string))
+		if err != nil {
+			return nil, fmt.Errorf("Error in Decode String for \n%v\n for schema %s: %v", value, schema.Name, err)
+		}
+		return []byte(v), nil
+
+	case yang.Ybool:
+		if strings.ToLower(value.(string)) == "false" {
+			return false, nil
+		} else if strings.ToLower(value.(string)) == "true" {
+			return true, nil
+		} else {
+			return nil, fmt.Errorf("Error in converting to yang bool type for the value for \n%v\n for schema %s", value, schema.Name)
+		}
+
+	case yang.Ystring:
+		return value.(string), nil
+
+	case yang.Ydecimal64:
+		floatV, err := strconv.ParseFloat(value.(string), 64)
+		if err != nil {
+			return nil, fmt.Errorf("Error parsing %v for schema %s: %v", value, schema.Name, err)
+		}
+		return floatV, nil
+
+	case yang.Yenum, yang.Yidentityref:
+		return enumStringToValue(parent, fieldName, value.(string))
+
+	case yang.Yint64:
+		// TODO(b/64812268): value types are different for internal style JSON.
+		intV, err := strconv.ParseInt(value.(string), 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("Error parsing %v for schema %s: %v", value, schema.Name, err)
+		}
+		return intV, nil
+
+	case yang.Yuint64:
+		uintV, err := strconv.ParseUint(value.(string), 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("Error parsing %v for schema %s: %v", value, schema.Name, err)
+		}
+		return uintV, nil
+
+	case yang.Yint8, yang.Yint16, yang.Yint32, yang.Yuint8, yang.Yuint16, yang.Yuint32:
+		gt := reflect.TypeOf(yangBuiltinTypeToGoType(ykind))
+		if rv, err := StringToType(gt, value.(string)); err != nil {
+			return nil, fmt.Errorf("Error StringToType:parsing %v for schema %s: %v", value, schema.Name, err)
+		} else {
+			return rv.Interface(), nil
+		}
+	}
+	return nil, fmt.Errorf("Error: unsupported type %v in schema node %s", ykind, schema.Name)
+}
+
 // sanitizeGNMI decodes the GNMI TypedValue encoded value into a field of the
 // corresponding type in GoStruct. Parent is the parent struct containing the
 // field being unmarshaled. schema is *yang.Entry corresponding to the field.
@@ -766,6 +845,18 @@ func sanitizeJSON(parent interface{}, schema *yang.Entry, fieldName string, valu
 func sanitizeGNMI(parent interface{}, schema *yang.Entry, fieldName string, tv *gpb.TypedValue, jsonTolerance bool) (interface{}, error) {
 	ykind := schema.Type.Kind
 
+	//get the original schema of the node and check whether it is leaf-list
+	if ygNode := schema.Parent.Dir[schema.Name]; ygNode != nil && ygNode.IsLeafList() {
+		// convert the leaf-list's value only if its of type string(TypedValue_StringVal)
+		if len(tv.GetStringVal()) > 0 {
+			if goVal, err := yangTypeToGoTypeVal(parent, schema, fieldName, tv.GetStringVal()); err == nil {
+				if gnmiTypeVal, err := value.FromScalar(goVal); err == nil {
+					tv = gnmiTypeVal
+				}
+			}
+		}
+	}
+
 	var ok bool
 	if ok = gNMIToYANGTypeMatches(ykind, tv, jsonTolerance); !ok {
 		return nil, fmt.Errorf("failed to unmarshal %v into %v", tv.GetValue(), yang.TypeKindToName[ykind])
@@ -815,6 +906,8 @@ func sanitizeGNMI(parent interface{}, schema *yang.Entry, fieldName string, tv *
 			return fv, nil
 		case *gpb.TypedValue_FloatVal:
 			return float64(v.FloatVal), nil
+		case *gpb.TypedValue_DoubleVal:
+			return float64(v.DoubleVal), nil
 		}
 	}
 	return nil, fmt.Errorf("%v type isn't expected for GNMIEncoding", yang.TypeKindToName[ykind])
@@ -848,9 +941,12 @@ func gNMIToYANGTypeMatches(ykind yang.TypeKind, tv *gpb.TypedValue, jsonToleranc
 	case yang.Ybinary:
 		_, ok = tv.GetValue().(*gpb.TypedValue_BytesVal)
 	case yang.Ydecimal64:
-		_, ok = tv.GetValue().(*gpb.TypedValue_DecimalVal)
+		_, ok = tv.GetValue().(*gpb.TypedValue_DoubleVal)
 		if !ok {
 			_, ok = tv.GetValue().(*gpb.TypedValue_FloatVal)
+			if !ok {
+				_, ok = tv.GetValue().(*gpb.TypedValue_DecimalVal)
+			}
 		}
 	}
 	return ok
diff --git a/ytypes/list.go b/ytypes/list.go
index 3244f8c..47de993 100644
--- a/ytypes/list.go
+++ b/ytypes/list.go
@@ -12,6 +12,9 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// This file is changed by Broadcom.
+// Modifications - Copyright 2019 Broadcom. The term Broadcom refers to Broadcom Inc. and/or its subsidiaries.
+
 package ytypes
 
 import (
@@ -19,6 +22,8 @@ import (
 	"reflect"
 	"strings"
 
+	log "github.com/golang/glog"
+	"github.com/google/go-cmp/cmp"
 	"github.com/kylelemons/godebug/pretty"
 	"github.com/openconfig/goyang/pkg/yang"
 	"github.com/openconfig/ygot/util"
@@ -217,6 +222,9 @@ func validateListSchema(schema *yang.Entry) error {
 		if len(schema.Key) == 0 {
 			return fmt.Errorf("list %s with config set must have a key", schema.Name)
 		}
+		if schema.PerfOpts.IsSchemaValidated.Load() {
+			return nil
+		}
 		keys := strings.Fields(schema.Key)
 		keysMissing := make(map[string]bool)
 		for _, v := range keys {
@@ -230,6 +238,7 @@ func validateListSchema(schema *yang.Entry) error {
 		}
 	}
 
+	schema.PerfOpts.IsSchemaValidated.Store(true)
 	return nil
 }
 
@@ -269,6 +278,31 @@ func nameMatchesPath(fieldName string, path []string) (bool, error) {
 	return false, fmt.Errorf("expected field %s path to have one or two elements, got %v", fieldName, path)
 }
 
+func isKeyHasStructPtr(key reflect.Value) bool {
+	switch key.Type().Kind() {
+	case reflect.Ptr:
+		return true
+	case reflect.Interface:
+		key = key.Elem()
+		switch key.Type().Kind() {
+		case reflect.Ptr:
+			return true
+		case reflect.Struct:
+		default:
+			return false
+		}
+	case reflect.Struct:
+	default:
+		return false
+	}
+	for i, n := 0, key.NumField(); i < n; i++ {
+		if isKeyHasStructPtr(key.Field(i)) {
+			return true
+		}
+	}
+	return false
+}
+
 // unmarshalList unmarshals a JSON array into a list parent, which must be a
 // map or slice ptr.
 //   schema is the schema of the schema node corresponding to the struct being
@@ -315,6 +349,23 @@ func unmarshalList(schema *yang.Entry, parent interface{}, jsonList interface{},
 		return fmt.Errorf("unmarshalList for %s parent type %T, has bad field type %v", listElementType, parent, listElementType)
 	}
 
+	isUpdate := (util.IsTypeMap(t) && isAllowUpdateInListMap(opts))
+	isListUpdate := isUpdate
+	mapV := reflect.ValueOf(parent)
+
+	isKeyHasPtr := false
+	keyCnt := 0
+	if isListUpdate {
+		keyCnt = mapV.Len()
+		if keyCnt == 0 {
+			isListUpdate = false
+		} else {
+			itr := mapV.MapRange()
+			itr.Next()
+			isKeyHasPtr = isKeyHasStructPtr(itr.Key())
+		}
+	}
+
 	// Iterate over JSON list. Each JSON list element is a map with the field
 	// name as the key. The JSON values must be unmarshaled and inserted into
 	// the new struct list element. When all fields of the new element have been
@@ -323,13 +374,32 @@ func unmarshalList(schema *yang.Entry, parent interface{}, jsonList interface{},
 	// types respectively.
 	// For a keyed list, the value(s) of the key are derived from the key fields
 	// in the new list element.
+	keyMap := make(map[string]bool)
 	for _, le := range jl {
 		var err error
 		jt := le.(map[string]interface{})
 		newVal := reflect.New(listElementType.Elem())
 		util.DbgPrint("creating a new list element val of type %v", newVal.Type())
-		if err := unmarshalStruct(schema, newVal.Interface(), jt, enc, opts...); err != nil {
-			return err
+		keyStr := ""
+		if isUpdate {
+			keyStr, err = unmarshalKeyFieldsInStruct(schema, parent, newVal.Interface(), jt, enc, opts...)
+			if err != nil {
+				log.Warning("Error in unmarshaling key fields: ", err, " and the structure type: ", newVal.Type())
+				pretty.Print(newVal.Interface())
+				pretty.Print(jt)
+				log.Info("Error in unmarshaling: schema key: ", schema.Key, " - parent type : ", reflect.TypeOf(parent), " - key type", reflect.TypeOf(parent).Key())
+				return err
+			} else if log.V(6) {
+				log.Info("unmarshalKeyFieldsInStruct: key in string format: ", keyStr)
+			}
+			if keyCnt == 0 && !keyMap[keyStr] {
+				isListUpdate = false
+			}
+		}
+		if !isListUpdate {
+			if err := unmarshalStruct(schema, newVal.Interface(), jt, enc, opts...); err != nil {
+				return err
+			}
 		}
 
 		switch {
@@ -339,7 +409,42 @@ func unmarshalList(schema *yang.Entry, parent interface{}, jsonList interface{},
 			if err != nil {
 				return err
 			}
-			err = util.InsertIntoMap(parent, newKey.Interface(), newVal.Interface())
+			isKeyExist := false
+			if isListUpdate {
+				if isKeyHasPtr {
+					iter := mapV.MapRange()
+					for iter.Next() {
+						if cmp.Equal(iter.Key().Interface(), newKey.Interface()) {
+							newVal = iter.Value()
+							isKeyExist = true
+							if !keyMap[keyStr] {
+								keyCnt--
+								keyMap[keyStr] = true
+							}
+							break
+						}
+					}
+				} else if listKV := mapV.MapIndex(reflect.ValueOf(newKey.Interface())); listKV.IsValid() {
+					if log.V(6) {
+						log.Info("Unmarshaling: key's type: ", newKey.Type(), " with value: ", newKey, " already exist in the map for the structure: ", listKV.Type())
+						log.Info("Unmarshaling: schema.Key: ", schema.Key, " - parent type : ", reflect.TypeOf(parent), " - key type", reflect.TypeOf(parent).Key())
+						pretty.Print(listKV.Interface())
+						pretty.Print(jt)
+					}
+					newVal = listKV
+					isKeyExist = true
+					if !keyMap[keyStr] {
+						keyCnt--
+						keyMap[keyStr] = true
+					}
+				}
+				if err := unmarshalStruct(schema, newVal.Interface(), jt, enc, opts...); err != nil {
+					return err
+				}
+			}
+			if !isKeyExist {
+				err = util.InsertIntoMap(parent, newKey.Interface(), newVal.Interface())
+			}
 		case util.IsTypeSlicePtr(t):
 			err = util.InsertIntoSlice(parent, newVal.Interface())
 		default:
@@ -349,7 +454,9 @@ func unmarshalList(schema *yang.Entry, parent interface{}, jsonList interface{},
 			return err
 		}
 	}
-	util.DbgPrint("list after unmarshal:\n%s\n", pretty.Sprint(parent))
+	if util.IsDebugLibraryEnabled() {
+		util.DbgPrint("list after unmarshal:\n%s\n", pretty.Sprint(parent))
+	}
 
 	return nil
 }
@@ -490,6 +597,8 @@ func makeKeyForInsert(schema *yang.Entry, parentMap interface{}, newVal reflect.
 	kv, err := getKeyValue(newVal.Elem(), schema.Key)
 	if err != nil {
 		return reflect.ValueOf(nil), err
+	} else if kv == nil {
+		return reflect.ValueOf(nil), fmt.Errorf("List key leaf node %v is not present in the payload", schema.Key)
 	}
 	util.DbgPrint("key value is %v.", kv)
 
@@ -516,6 +625,9 @@ func insertAndGetKey(schema *yang.Entry, root interface{}, keys map[string]strin
 	}
 
 	// TODO(yusufsn): When the key is a leafref, its target should be filled out.
+	if len(keys) == 0 {
+		return nil, nil
+	}
 	mapVal, err := makeValForInsert(schema, root, keys)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create map value for insert, root %T, keys %v: %v", root, keys, err)
@@ -579,3 +691,43 @@ func getKeyValue(structVal reflect.Value, key string) (interface{}, error) {
 
 	return nil, fmt.Errorf("could not find key field %s in struct type %s", key, structVal.Type())
 }
+
+// UnmarshalListKey creates the key and its value (ygot struct) and insert into the map return the created ygot struct
+// if the key is not already present in the map
+func UnmarshalListKey(schema *yang.Entry, root interface{}, keys map[string]string) (interface{}, error) {
+	switch {
+	case schema.Key == "":
+		return nil, fmt.Errorf("unkeyed list can't be traversed, type %T, keys %v", root, keys)
+	case !util.IsValueMap(reflect.ValueOf(root)):
+		return nil, fmt.Errorf("root has type %T, want map", root)
+	}
+	if len(keys) == 0 {
+		return nil, nil
+	}
+
+	mapVal, err := makeValForInsert(schema, root, keys)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create map value for insert, root %T, keys %v: %v", root, keys, err)
+	}
+
+	mapKey, err := makeKeyForInsert(schema, root, mapVal)
+	if err != nil {
+		return nil, err
+	}
+
+	mapV := reflect.ValueOf(root)
+
+	if listKV := mapV.MapIndex(mapKey); listKV.IsValid() {
+		if log.V(5) {
+			log.Info("Unmarshaling: key's type: ", mapKey.Type(), " with value: ", mapKey, " already exist in the map for the structure: ", listKV.Type())
+		}
+		return listKV.Interface(), nil
+	}
+
+	err = util.InsertIntoMap(root, mapKey.Interface(), mapVal.Interface())
+	if err != nil {
+		return nil, fmt.Errorf("failed to insert into map %T, keys %v: %v", root, keys, err)
+	}
+
+	return mapVal.Interface(), nil
+}
diff --git a/ytypes/node.go b/ytypes/node.go
index 4baa3ac..9d7139a 100644
--- a/ytypes/node.go
+++ b/ytypes/node.go
@@ -12,10 +12,14 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// This file is changed by Broadcom.
+// Modifications - Copyright 2019 Broadcom. The term Broadcom refers to Broadcom Inc. and/or its subsidiaries.
+
 package ytypes
 
 import (
 	"reflect"
+	"strings"
 
 	"github.com/openconfig/goyang/pkg/yang"
 	"github.com/openconfig/ygot/util"
@@ -56,6 +60,9 @@ type retrieveNodeArgs struct {
 	// specifically to deal with uint values being streamed as positive int
 	// values.
 	tolerateJSONInconsistenciesForVal bool
+	// if retListMap is set to true then target list map will be retrieved
+	// if the target node of the path is list without keys
+	retListMap bool
 	// preferShadowPath uses the name of the "shadow-path" tag of a
 	// GoStruct to determine the path elements instead of the
 	// "path" tag, whenever the former is present.
@@ -165,6 +172,16 @@ func retrieveNodeContainer(schema *yang.Entry, root interface{}, path *gpb.Path,
 				if err := util.InitializeStructField(root, ft.Name, args.initializeLeafs); err != nil {
 					return nil, status.Errorf(codes.Unknown, "failed to initialize struct field %s in %T, child schema %v, path %v", ft.Name, root, cschema, path)
 				}
+
+				if cschema.IsLeaf() || cschema.IsLeafList() {
+					if len(path.Elem) == 1 && len(path.Elem[0].Key) == 1 {
+						var vals []string
+						vals = append(vals, path.Elem[0].Key[path.Elem[0].Name])
+						if args.val, err = ygot.EncodeTypedValue(vals, gpb.Encoding_JSON_IETF); err != nil {
+							return nil, status.Errorf(codes.Unknown, "failed to get the typed value '%v' for leaf/leaf-list => %s in %T ; because of %v", vals, ft.Name, root, err)
+						}
+					}
+				}
 			}
 
 			// If delete is specified, and the path is exhausted, then we set the
@@ -265,6 +282,11 @@ func retrieveNodeList(schema *yang.Entry, root interface{}, path, traversedPath
 		return nil, status.Errorf(codes.InvalidArgument, "root has type %T, expect map", root)
 	}
 
+	if len(path.GetElem()[0].GetKey()) == 0 && args.retListMap {
+		// returning list map, since there is no keys present in the path
+		return []*TreeNode{{Path: traversedPath, Schema: schema, Data: root}}, nil
+	}
+
 	var matches []*TreeNode
 
 	listKeyT := rv.Type().Key()
@@ -306,6 +328,17 @@ func retrieveNodeList(schema *yang.Entry, root interface{}, path, traversedPath
 			if err != nil {
 				return nil, status.Errorf(codes.InvalidArgument, "failed to convert %v to a string, path %v: %v", kv, path, err)
 			}
+
+			if strings.Contains(keyAsString, "\\") {
+				keyPath, err := ygot.StringToStructuredPath(schema.Name + "[" + schema.Key + "=" + keyAsString + "]")
+				if err != nil {
+					return nil, status.Errorf(codes.InvalidArgument, "failed to convert %v to a string, path %v: %v", kv, path, err)
+				}
+				if len(keyPath.Elem) == 1 {
+					keyAsString = keyPath.Elem[0].Key[schema.Key]
+				}
+			}
+
 			if keyAsString == pathKey {
 				remainingPath := util.PopGNMIPath(path)
 				if args.delete && len(remainingPath.GetElem()) == 0 {
@@ -350,6 +383,15 @@ func retrieveNodeList(schema *yang.Entry, root interface{}, path, traversedPath
 			if err != nil {
 				return nil, status.Errorf(codes.Unknown, "failed to convert the field value to string, field %v: %v", fieldName, err)
 			}
+			if strings.Contains(keyAsString, "\\") {
+				keyPath, err := ygot.StringToStructuredPath(schema.Name + "[" + schemaKey + "=" + keyAsString + "]")
+				if err != nil {
+					return nil, status.Errorf(codes.Unknown, "failed to convert the field value to string, field %v: %v", fieldName, err)
+				}
+				if len(keyPath.Elem) == 1 {
+					keyAsString = keyPath.Elem[0].Key[schemaKey]
+				}
+			}
 			if !(args.handleWildcards && pathKey == "*") && pathKey != keyAsString {
 				match = false
 				break
@@ -382,6 +424,11 @@ func retrieveNodeList(schema *yang.Entry, root interface{}, path, traversedPath
 		if err != nil {
 			return nil, err
 		}
+
+		if key == nil {
+			return []*TreeNode{{Path: traversedPath, Schema: schema, Data: root}}, nil
+		}
+
 		nodes, err := retrieveNode(schema, rv.MapIndex(reflect.ValueOf(key)).Interface(), util.PopGNMIPath(path), appendElem(traversedPath, path.GetElem()[0]), args)
 		if err != nil {
 			return nil, err
@@ -422,6 +469,18 @@ func GetOrCreateNode(schema *yang.Entry, root interface{}, path *gpb.Path, opts
 	return nodes[0].Data, nodes[0].Schema, nil
 }
 
+// GetOrCreateTargetNode function is similar to GetOrCreateNode function, but it returns the list map object
+// if the target node of the path is list without keys
+func GetOrCreateTargetNode(schema *yang.Entry, root interface{}, path *gpb.Path) (interface{}, *yang.Entry, error) {
+	nodes, err := retrieveNode(schema, root, path, nil, retrieveNodeArgs{modifyRoot: true, retListMap: true})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// There must be a result as this function initializes nodes along the supplied path.
+	return nodes[0].Data, nodes[0].Schema, nil
+}
+
 // TreeNode wraps an individual entry within a YANG data tree to return to a caller.
 type TreeNode struct {
 	// Schema is the schema entry for the data tree node, specified as a goyang Entry struct.
diff --git a/ytypes/string_type.go b/ytypes/string_type.go
index 93639ee..670f3d1 100644
--- a/ytypes/string_type.go
+++ b/ytypes/string_type.go
@@ -12,6 +12,9 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// This file is changed by Broadcom.
+// Modifications - Copyright 2019 Broadcom. The term Broadcom refers to Broadcom Inc. and/or its subsidiaries.
+
 package ytypes
 
 import (
@@ -189,6 +192,12 @@ func validateStringSchema(schema *yang.Entry) error {
 		return fmt.Errorf("string schema %s has wrong type %v", schema.Name, schema.Type.Kind)
 	}
 
+	if schema.PerfOpts.IsSchemaValidated.Load() {
+		return nil
+	}
+
+	var err error
+
 	patterns, isPOSIX := util.SanitizedPattern(schema.Type)
 	for _, p := range patterns {
 		if _, err := reCache.compilePattern(p, isPOSIX); err != nil {
@@ -196,5 +205,8 @@ func validateStringSchema(schema *yang.Entry) error {
 		}
 	}
 
-	return validateLengthSchema(schema)
+	if err = validateLengthSchema(schema); err == nil {
+		schema.PerfOpts.IsSchemaValidated.Store(true)
+	}
+	return err
 }
diff --git a/ytypes/unmarshal.go b/ytypes/unmarshal.go
index b28e8e5..b83577f 100644
--- a/ytypes/unmarshal.go
+++ b/ytypes/unmarshal.go
@@ -12,6 +12,9 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// This file is changed by Broadcom.
+// Modifications - Copyright 2019 Broadcom. The term Broadcom refers to Broadcom Inc. and/or its subsidiaries.
+
 package ytypes
 
 import (
@@ -39,6 +42,13 @@ type IgnoreExtraFields struct{}
 // IsUnmarshalOpt marks IgnoreExtraFields as a valid UnmarshalOpt.
 func (*IgnoreExtraFields) IsUnmarshalOpt() {}
 
+// AllowUpdateInListMap is an unmarshal option to update the key and value
+// if it is present in the list map already
+type AllowUpdateInListMap struct{}
+
+// IsUnmarshalOpt marks AllowUpdateInListMap as a valid UnmarshalOpt.
+func (*AllowUpdateInListMap) IsUnmarshalOpt() {}
+
 // Unmarshal recursively unmarshals JSON data tree in value into the given
 // parent, using the given schema. Any values already in the parent that are
 // not present in value are preserved. If provided schema is a leaf or leaf
@@ -77,7 +87,9 @@ func unmarshalGeneric(schema *yang.Entry, parent interface{}, value interface{},
 	if schema == nil {
 		return fmt.Errorf("nil schema for parent type %T, value %v (%T)", parent, value, value)
 	}
-	util.DbgPrint("Unmarshal value %v, type %T, into parent type %T, schema name %s", util.ValueStrDebug(value), value, parent, schema.Name)
+	if util.IsDebugLibraryEnabled() {
+		util.DbgPrint("Unmarshal value %v, type %T, into parent type %T, schema name %s", util.ValueStrDebug(value), value, parent, schema.Name)
+	}
 
 	if enc == GNMIEncoding && !(schema.IsLeaf() || schema.IsLeafList()) {
 		return errors.New("unmarshalling a non leaf node isn't supported in GNMIEncoding mode")
@@ -108,3 +120,14 @@ func hasIgnoreExtraFields(opts []UnmarshalOpt) bool {
 	}
 	return false
 }
+
+// isAllowUpdateInListMap determines whether the supplied slice of UnmarshalOpts contains
+// the AllowUpdateInListMap option.
+func isAllowUpdateInListMap(opts []UnmarshalOpt) bool {
+	for _, o := range opts {
+		if _, ok := o.(*AllowUpdateInListMap); ok {
+			return true
+		}
+	}
+	return false
+}
diff --git a/ytypes/util_types.go b/ytypes/util_types.go
index 33c6edb..d6f95a7 100644
--- a/ytypes/util_types.go
+++ b/ytypes/util_types.go
@@ -457,8 +457,9 @@ func checkJSONFloat64Range(t yang.TypeKind, f float64) error {
 // yangTypeToLeafEntry returns a leaf Entry with Type set to t.
 func yangTypeToLeafEntry(t *yang.YangType) *yang.Entry {
 	return &yang.Entry{
-		Kind: yang.LeafEntry,
-		Type: t,
+		Kind:     yang.LeafEntry,
+		Type:     t,
+		PerfOpts: &yang.PerformanceOpts{},
 	}
 }
 
@@ -469,5 +470,6 @@ func yangKindToLeafEntry(k yang.TypeKind) *yang.Entry {
 		Type: &yang.YangType{
 			Kind: k,
 		},
+		PerfOpts: &yang.PerformanceOpts{},
 	}
 }
