diff --git a/gnmi_get/gnmi_get.go b/gnmi_get/gnmi_get.go
index dbf737b..96e6443 100644
--- a/gnmi_get/gnmi_get.go
+++ b/gnmi_get/gnmi_get.go
@@ -19,6 +19,7 @@ package main
 import (
 	"flag"
 	"fmt"
+	"io/ioutil"
 	"strings"
 	"time"
 
@@ -67,12 +68,16 @@ var (
 	targetName       = flag.String("target_name", "hostname.com", "The target name use to verify the hostname returned by TLS handshake")
 	timeOut          = flag.Duration("time_out", 10*time.Second, "Timeout for the Get request, 10 seconds by default")
 	encodingName     = flag.String("encoding", "JSON_IETF", "value encoding format to be used")
+	jwtToken         = flag.String("jwt_token", "", "JWT Token if required")
+	pathTarget       = flag.String("xpath_target", "", "name of the target for which the path is a member")
+	protoFileFlags   arrayFlags
 )
 
 func main() {
 	flag.Var(&xPathFlags, "xpath", "xpath of the config node to be fetched")
 	flag.Var(&pbPathFlags, "pbpath", "protobuf format path of the config node to be fetched")
 	flag.Var(&pbModelDataFlags, "model_data", "Data models to be used by the target in the format of 'name,organization,version'")
+	flag.Var(&protoFileFlags, "proto_file", "Files for proto bytes in get response")
 	flag.Parse()
 
 	opts := credentials.ClientCredentials(*targetName)
@@ -86,6 +91,9 @@ func main() {
 
 	ctx, cancel := context.WithTimeout(context.Background(), *timeOut)
 	defer cancel()
+	if len(*jwtToken) > 0 {
+		ctx = credentials.AttachJWT(ctx, *jwtToken)
+	}
 
 	encoding, ok := pb.Encoding_value[*encodingName]
 	if !ok {
@@ -119,7 +127,11 @@ func main() {
 		}
 	}
 
+	var prefix pb.Path
+	prefix.Target = *pathTarget
+
 	getRequest := &pb.GetRequest{
+		Prefix: &prefix,
 		Encoding:  pb.Encoding(encoding),
 		Path:      pbPathList,
 		UseModels: pbModelDataList,
@@ -135,4 +147,20 @@ func main() {
 
 	fmt.Println("== getResponse:")
 	utils.PrintProto(getResponse)
+
+	if *encodingName == "PROTO" {
+		cnt := 0
+		for _, notification := range getResponse.GetNotification() {
+			for _, update := range notification.GetUpdate() {
+				val := update.GetVal()
+				protoBytes := val.GetProtoBytes()
+				fileName := protoFileFlags[cnt]
+				err := ioutil.WriteFile(fileName, protoBytes, 0666)
+				if err != nil {
+					log.Exitf("Write PROTO %v failed", cnt)
+				}
+				cnt = cnt + 1
+			}
+		}
+	}
 }
diff --git a/gnmi_set/gnmi_set.go b/gnmi_set/gnmi_set.go
index c5af997..459b9ce 100644
--- a/gnmi_set/gnmi_set.go
+++ b/gnmi_set/gnmi_set.go
@@ -54,13 +54,30 @@ var (
 	targetAddr = flag.String("target_addr", "localhost:10161", "The target address in the format of host:port")
 	targetName = flag.String("target_name", "hostname.com", "The target name use to verify the hostname returned by TLS handshake")
 	timeOut    = flag.Duration("time_out", 10*time.Second, "Timeout for the Get request, 10 seconds by default")
+	pathTarget = flag.String("xpath_target", "", "name of the target for which the path is a member")
+	jwtToken   = flag.String("jwt_token", "", "JWT Token if required")
 )
 
 func buildPbUpdateList(pathValuePairs []string) []*pb.Update {
 	var pbUpdateList []*pb.Update
 	for _, item := range pathValuePairs {
-		pathValuePair := strings.SplitN(item, ":", 2)
+		pathValuePair := make([]string, 2)
 		// TODO (leguo): check if any path attribute contains ':'
+		modName := strings.SplitN(item, "/", 3)
+		lc := strings.LastIndex(modName[2],":")
+
+		if lc == -1 {
+			log.Exitf("invalid path-value pair: %v", item)
+		}
+		pathValuePair_r := strings.SplitN(modName[2], ":", 2)
+		// pathValuePair[0] = modName[2][0:lc]
+
+		pathValuePair[0] = "/" + modName[1] + "/" + pathValuePair_r[0]
+
+		pathValuePair[1] = modName[2][lc+1:]
+		fmt.Println(pathValuePair[0])
+		fmt.Println(pathValuePair[1])
+
 		if len(pathValuePair) != 2 || len(pathValuePair[1]) == 0 {
 			log.Exitf("invalid path-value pair: %v", item)
 		}
@@ -69,7 +86,9 @@ func buildPbUpdateList(pathValuePairs []string) []*pb.Update {
 			log.Exitf("error in parsing xpath %q to gnmi path", pathValuePair[0])
 		}
 		var pbVal *pb.TypedValue
-		if pathValuePair[1][0] == '@' {
+		if pathValuePair[1][0] == '#' {
+			pbVal = nil
+		} else if pathValuePair[1][0] == '@' {
 			jsonFile := pathValuePair[1][1:]
 			jsonConfig, err := ioutil.ReadFile(jsonFile)
 			if err != nil {
@@ -81,6 +100,17 @@ func buildPbUpdateList(pathValuePairs []string) []*pb.Update {
 					JsonIetfVal: jsonConfig,
 				},
 			}
+		} else if pathValuePair[1][0] == '$' {
+			protoFile := pathValuePair[1][1:]
+			protoConfig, err := ioutil.ReadFile(protoFile)
+			if err != nil {
+				log.Exitf("cannot read data from file %v", protoFile)
+			}
+			pbVal = &pb.TypedValue{
+				Value: &pb.TypedValue_ProtoBytes{
+					ProtoBytes: protoConfig,
+				},
+			}
 		} else {
 			if strVal, err := strconv.Unquote(pathValuePair[1]); err == nil {
 				pbVal = &pb.TypedValue{
@@ -145,7 +175,10 @@ func main() {
 	replaceList := buildPbUpdateList(replaceOpt)
 	updateList := buildPbUpdateList(updateOpt)
 
+	var prefix pb.Path
+	prefix.Target = *pathTarget
 	setRequest := &pb.SetRequest{
+		Prefix:    &prefix,
 		Delete:  deleteList,
 		Replace: replaceList,
 		Update:  updateList,
@@ -155,11 +188,17 @@ func main() {
 	utils.PrintProto(setRequest)
 
 	cli := pb.NewGNMIClient(conn)
-	setResponse, err := cli.Set(context.Background(), setRequest)
+	ctx, cancel := context.WithTimeout(context.Background(), *timeOut)
+	defer cancel()
+
+	if len(*jwtToken) > 0 {
+		ctx = credentials.AttachJWT(ctx, *jwtToken)
+	}
+	setResponse, err := cli.Set(ctx, setRequest)
 	if err != nil {
 		log.Exitf("Set failed: %v", err)
 	}
 
-	fmt.Println("== getResponse:")
+	fmt.Println("== setResponse:")
 	utils.PrintProto(setResponse)
 }
diff --git a/utils/credentials/credentials.go b/utils/credentials/credentials.go
index 0fea80c..73c9d5b 100644
--- a/utils/credentials/credentials.go
+++ b/utils/credentials/credentials.go
@@ -159,3 +159,7 @@ func AuthorizeUser(ctx context.Context) (string, bool) {
 	}
 	return fmt.Sprintf("not authorized with \"%s:%s\"", user[0], pass[0]), false
 }
+
+func AttachJWT(ctx context.Context, token string) context.Context {
+	return metadata.AppendToOutgoingContext(ctx, "access_token", token)
+}
diff --git a/utils/xpath/path.go b/utils/xpath/path.go
index 57f0538..5994d4e 100644
--- a/utils/xpath/path.go
+++ b/utils/xpath/path.go
@@ -18,6 +18,7 @@ package xpath
 
 import (
 	"fmt"
+	"strings"
 
 	pb "github.com/openconfig/gnmi/proto/gnmi"
 )
@@ -45,9 +46,17 @@ func ToGNMIPath(xpath string) (*pb.Path, error) {
 		return nil, err
 	}
 	var pbPathElements []*pb.PathElem
-	for _, elem := range xpathElements {
+	origin := ""
+	for i, elem := range xpathElements {
 		switch v := elem.(type) {
 		case string:
+			if i == 0 {
+				firstElement := strings.SplitN(v, ":", 2)
+				if len(firstElement) == 2 {
+					origin = firstElement[0]
+					v = firstElement[1]
+				}
+			}
 			pbPathElements = append(pbPathElements, &pb.PathElem{Name: v})
 		case map[string]string:
 			n := len(pbPathElements)
@@ -62,5 +71,5 @@ func ToGNMIPath(xpath string) (*pb.Path, error) {
 			return nil, fmt.Errorf("wrong data type: %T", v)
 		}
 	}
-	return &pb.Path{Elem: pbPathElements}, nil
+	return &pb.Path{Origin: origin, Elem: pbPathElements}, nil
 }
diff --git a/utils/xpath/xpath.go b/utils/xpath/xpath.go
index 9db1af1..aa40918 100644
--- a/utils/xpath/xpath.go
+++ b/utils/xpath/xpath.go
@@ -88,17 +88,19 @@ func splitPath(str string) ([]string, error) {
 	for end < len(str) {
 		switch str[end] {
 		case '/':
-			if !insideBrackets {
+			if (end == 0 || str[end-1] != '\\') && !insideBrackets {
 				// Current '/' is a valid path element
 				// separator.
 				if end > begin {
-					path = append(path, str[begin:end])
+					// Unescape any '/' in the path element
+					tmp := strings.Replace(str[begin:end], "\\/", "/", -1)
+					path = append(path, tmp)
 				}
 				end++
 				begin = end
 			} else {
 				// Current '/' must be part of a List key value
-				// string.
+				// string or is an escaped '/' in a path element.
 				end++
 			}
 		case '[':
@@ -211,7 +213,7 @@ func parseElement(elem string) ([]interface{}, error) {
 	i := strings.Index(elem, "[")
 	if i < 0 {
 		if !idRe.MatchString(elem) {
-			return nil, fmt.Errorf("invalid node name: %q", elem)
+			// return nil, fmt.Errorf("invalid node name: %q", elem)
 		}
 		return []interface{}{elem}, nil
 	}
