diff --git a/BUILD.bazel b/BUILD.bazel
new file mode 100644
--- /dev/null
+++ b/BUILD.bazel
@@ -0,0 +1,22 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_library")
+
+go_library(
+    name = "pam",
+    srcs = [
+        "callback.go",
+        "transaction.c",
+        "transaction.go",
+    ],
+    cgo = True,
+    cdeps = [
+        "@bookworm//libpam0g-dev:libpam0g",
+    ],
+    cppopts = [
+        "-I$(BINDIR)/external/rules_distroless++apt+bookworm_libpam0g-dev-amd64_1.5.2-6-deb12u1/usr/include/security",
+        "-I$(BINDIR)/external/rules_distroless++apt+bookworm_libpam0g-dev-amd64_1.5.2-6-deb12u1/usr/include",
+    ],
+    copts = ["-Wall", "-std=c99"],
+    clinkopts = ["-ldl"],
+    importpath = "github.com/msteinert/pam",
+    visibility = ["//visibility:public"],
+)
