diff --git a/cli/cli.go b/cli/cli.go
index 0d3c4d6..26f98f7 100644
--- a/cli/cli.go
+++ b/cli/cli.go
@@ -20,8 +20,10 @@ package cli
 import (
 	"bytes"
 	"context"
+	"encoding/json"
 	"errors"
 	"fmt"
+	"reflect"
 	"sort"
 	"strings"
 	"time"
@@ -294,7 +296,8 @@ func displayPollingResults(ctx context.Context, query client.Query, cfg *Config)
 // subsequent individual updates as they arrive.
 func displayStreamingResults(ctx context.Context, query client.Query, cfg *Config) error {
 	c := client.New()
-	complete := false
+	onChange := query.Streaming_type == gpb.SubscriptionMode(1)
+	complete := onChange
 	display := func(path []string, ts time.Time, val interface{}) {
 		if !complete {
 			return
@@ -303,8 +306,15 @@ func displayStreamingResults(ctx context.Context, query client.Query, cfg *Confi
 		if t := formatTime(ts, cfg); t != nil {
 			b.add(append(path, "timestamp"), t)
 			b.add(append(path, "value"), val)
-		} else {
+		} else if !onChange {
 			b.add(path, val)
+		} else {
+			v := reflect.ValueOf(val)
+			if v.Kind() == reflect.String {
+				json.Unmarshal([]byte(v.String()), &b)
+			} else {
+				b.add(path, val)
+			}
 		}
 		cfg.Display(b.display(cfg.DisplayPrefix, cfg.DisplayIndent))
 	}
diff --git a/client/gnmi/client.go b/client/gnmi/client.go
index 7bfe3b2..8b0cc8e 100644
--- a/client/gnmi/client.go
+++ b/client/gnmi/client.go
@@ -204,7 +204,7 @@ func (c *Client) defaultRecv(msg proto.Message) error {
 		}
 	case *gpb.SubscribeResponse_Update:
 		n := v.Update
-		p := path.ToStrings(n.Prefix, true)
+		p := path.ToStrings(n.Prefix, false) // patched for legacy tests
 		ts := time.Unix(0, n.Timestamp)
 		for _, u := range n.Update {
 			if u.Path == nil {
@@ -259,6 +259,7 @@ func subscribe(q client.Query) (*gpb.SubscribeRequest, error) {
 		Subscribe: &gpb.SubscriptionList{
 			Mode:   getType(q.Type),
 			Prefix: &gpb.Path{Target: q.Target},
+			Encoding: q.Encoding,
 		},
 	}
 	if q.UpdatesOnly {
@@ -269,7 +270,7 @@ func subscribe(q client.Query) (*gpb.SubscribeRequest, error) {
 		if err != nil {
 			return nil, fmt.Errorf("invalid query path %q: %v", qq, err)
 		}
-		s.Subscribe.Subscription = append(s.Subscribe.Subscription, &gpb.Subscription{Path: pp})
+		s.Subscribe.Subscription = append(s.Subscribe.Subscription, &gpb.Subscription{Path: pp, Mode: q.Streaming_type, SampleInterval: q.Streaming_sample_int, HeartbeatInterval: q.Heartbeat_int, SuppressRedundant: q.Suppress_redundant})
 	}
 	s.Subscribe.Encoding = q.Encoding
 	return &gpb.SubscribeRequest{Request: s}, nil
diff --git a/client/query.go b/client/query.go
index b36b27f..9ca9396 100644
--- a/client/query.go
+++ b/client/query.go
@@ -184,6 +184,11 @@ type Query struct {
 	TunnelConn *tunnel.Conn
 	// Encoding is the subscription response encoding value.
 	Encoding gpb.Encoding
+
+	Streaming_type       gpb.SubscriptionMode
+	Streaming_sample_int uint64
+	Heartbeat_int        uint64
+	Suppress_redundant   bool
 }

 // Destination extracts a Destination instance out of Query fields.
diff --git a/cmd/gnmi_cli/gnmi_cli.go b/cmd/gnmi_cli/gnmi_cli.go
index d9065e6..01a435e 100644
--- a/cmd/gnmi_cli/gnmi_cli.go
+++ b/cmd/gnmi_cli/gnmi_cli.go
@@ -28,6 +28,7 @@ import (
 	"context"
 	"crypto/tls"
 	"crypto/x509"
+	"encoding/json"
 	"errors"
 	"fmt"
 	"io/ioutil"
@@ -52,12 +53,31 @@ import (
 )

 var (
+	displayHandle = os.Stdout
+	prefix = []byte("[\n")
+	rcvd_cnt uint = 0
+	term = make(chan string, 1)
 	q   = client.Query{TLS: &tls.Config{}}
 	mu  sync.Mutex
 	cfg = cli.Config{Display: func(b []byte) {
-		defer mu.Unlock()
-		mu.Lock()
-		os.Stdout.Write(append(b, '\n'))
+		found := len(*expected_event) == 0
+		if !found {
+			var fvp map[string]interface{}
+
+			json.Unmarshal(b, &fvp)
+			_, found = fvp[*expected_event]
+		}
+		if found {
+			defer mu.Unlock()
+			mu.Lock()
+
+			if *expected_cnt > 0 {
+				rcvd_cnt += 1
+			}
+			displayHandle.Write(prefix)
+			displayHandle.Write(b)
+			prefix = []byte(",\n")
+		}
 	}}

 	clientTypes = flags.NewStringList(&cfg.ClientTypes, []string{gclient.Type})
@@ -80,6 +100,17 @@ var (
 	caCert       = flag.String("ca_crt", "", "CA certificate file. Used to verify server TLS certificate.")
 	clientCert   = flag.String("client_crt", "", "Client certificate file. Used for client certificate-based authentication.")
 	clientKey    = flag.String("client_key", "", "Client private key file. Used for client certificate-based authentication.")
+
+	//Subscribe Options
+	streaming_type       = flag.String("streaming_type", "TARGET_DEFINED", "One of TARGET_DEFINED, ON_CHANGE or SAMPLE")
+	streaming_sample_int = flag.Uint("streaming_sample_interval", 0, "Streaming sample inteval seconds, 0 means lowest supported.")
+	heartbeat_int        = flag.Uint("heartbeat_interval", 0, "Heartbeat inteval seconds.")
+	suppress_redundant   = flag.Bool("suppress_redundant", false, "Suppress Redundant Subscription Updates")
+	encoding             = flag.String("encoding", "JSON", "One of JSON, BYTES, PROTO, ASCII, JSON_IETF")
+	output_file          = flag.String("output_file", "", "Output file to write the response")
+	expected_cnt         = flag.Uint("expected_count", 0, "End upon receiving the count of responses.")
+	expected_event       = flag.String("expected_event", "", "Event to capture")
+	streaming_timeout    = flag.Uint("streaming_timeout", 0, "Exits after this time.")
 )

 func init() {
@@ -124,6 +155,20 @@ func init() {
 func main() {
 	flag.Parse()

+	defer func() {
+		displayHandle.Write([]byte("\n]\n"))
+		displayHandle.Close()
+	}()
+
+	if len(*output_file) != 0 {
+		var err error
+		displayHandle, err = os.OpenFile(*output_file, os.O_TRUNC|os.O_CREATE|os.O_WRONLY, 0644)
+		if err != nil {
+			log.Error(fmt.Printf("unable to create output file(%v) err=%v\n", *output_file, err))
+			return
+		}
+	}
+
 	ctx, cancel := context.WithCancel(context.Background())
 	// Terminate immediately on Ctrl+C, skipping lame-duck mode.
 	go func() {
@@ -133,6 +178,32 @@ func main() {
 		cancel()
 	}()

+	go func() {
+		if *streaming_timeout > 0 {
+			var sleep_cnt uint = 0
+			for sleep_cnt < *streaming_timeout {
+				time.Sleep(time.Second)
+				sleep_cnt += 1
+				if *expected_cnt <= rcvd_cnt {
+					s := fmt.Sprintf("Received all. expected:%d rcvd:%d", *expected_cnt, rcvd_cnt)
+					log.V(7).Infof("Writing to terminate: %v", s)
+					term <- s
+					return
+				}
+			}
+			s := fmt.Sprintf("Timeout %d Secs", *streaming_timeout)
+			log.V(7).Infof("Writing to terminate: %v", s)
+			term <- s
+		}
+	}()
+
+	go func() {
+		// Terminate when indicated.
+		m := <-term
+		log.V(1).Infof("Terminating due to %v", m)
+		cancel()
+	}()
+
 	if len(q.Addrs) == 0 {
 		log.Exit("--address must be set")
 	}
@@ -306,6 +377,27 @@ func executeSubscribe(ctx context.Context) error {
 	if len(*queryFlag) == 0 {
 		return errors.New("--query must be set")
 	}
+	if *streaming_type == "TARGET_DEFINED" {
+		q.Streaming_type = gpb.SubscriptionMode(0)
+	} else if *streaming_type == "ON_CHANGE" {
+		q.Streaming_type = gpb.SubscriptionMode(1)
+	} else if *streaming_type == "SAMPLE" {
+		q.Streaming_type = gpb.SubscriptionMode(2)
+	} else {
+		return errors.New("-streaming_type must be one of: (TARGET_DEFINED, ON_CHANGE, SAMPLE)")
+	}
+	q.Streaming_sample_int = uint64(*streaming_sample_int) * uint64(time.Second)
+	if *queryType == "streaming" || *queryType == "s" {
+		q.Heartbeat_int = uint64(*heartbeat_int) * uint64(time.Second)
+	} else if *heartbeat_int > 0 {
+		return errors.New("-heartbeat_interval only valid with streaming query type")
+	}
+	if enc, ok := gpb.Encoding_value[strings.ToUpper(*encoding)]; ok {
+		q.Encoding = gpb.Encoding(enc)
+	} else {
+		return fmt.Errorf("Invalid -encoding value \"%s\"", *encoding)
+	}
+	q.Suppress_redundant = bool(*suppress_redundant)
 	for _, path := range *queryFlag {
 		query, err := parseQuery(path, cfg.Delimiter)
 		if err != nil {
diff --git a/value/value.go b/value/value.go
index fd1149c..a34cee1 100644
--- a/value/value.go
+++ b/value/value.go
@@ -133,21 +133,13 @@ func ToScalar(tv *pb.TypedValue) (interface{}, error) {
 		if err := json.Unmarshal(v, &i); err != nil {
 			return nil, err
 		}
-		uVal := DeprecatedScalar{
-			Message: "Deprecated TypedValue_JsonVal",
-			Value:   i,
-		}
-		return uVal, nil
+		return i, nil  // Json Values are not yet deprecated and tests rely on the message
 	case *gpb.TypedValue_JsonIetfVal:
 		v := tv.GetJsonIetfVal()
 		if err := json.Unmarshal(v, &i); err != nil {
 			return nil, err
 		}
-		uVal := DeprecatedScalar{
-			Message: "Deprecated TypedValue_JsonIetfVal",
-			Value:   i,
-		}
-		return uVal, nil
+		return i, nil  // Json Values are not yet deprecated and tests rely on the message
 	default:
 		return nil, fmt.Errorf("non-scalar type %+v", tv.Value)
 	}
